{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational analysis to determine protein degradation motifs\n",
    "Description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "from scipy import stats \n",
    "import sys\n",
    "import sklearn\n",
    "from sklearn.preprocessing import normalize\n",
    "from sklearn.metrics import roc_curve, auc, roc_auc_score\n",
    "import pylab as pl\n",
    "from matplotlib import collections  as mc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "sys.path.insert(0, '/Users/cynthiachen/Downloads/Internship2019/degron_mutation/deepDegron')\n",
    "from train_nn import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "wd = '/Users/cynthiachen/Downloads/Internship2019/degron_mutation/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load pre-trained neural network models from pickle file\n",
    "model_bag_of_words = pickle.load(open( \"../models/neural_network_bag_of_words_v3.pickle\", \"rb\" ))\n",
    "model_pos_specific = pickle.load(open( \"../models/neural_network_pos_specific_v3.pickle\", \"rb\" ))\n",
    "\n",
    "# Create pd dataframe of sequence information \n",
    "degron_pred = pd.read_csv(\"../data/degron_pred.csv\", index_col=0)\n",
    "# Select column of input sequences\n",
    "input_seq = degron_pred.iloc[:, 10]\n",
    "\n",
    "\n",
    "# Use compute_feature_matrix function to encode sequences\n",
    "features = compute_feature_matrix(input_seq, 6, True)\n",
    "condensed_features = features[:, 0:20] # Remove zeros at the end for bag of words prediction\n",
    "\n",
    "# Use pre-trained model to predict sequence degredataion\n",
    "pred_bow_wt = model_bag_of_words.predict_proba(condensed_features) # wild-type bag of words prediction\n",
    "pred_ps_wt = model_pos_specific.predict_proba(features)            # wild-type position-specific prediction\n",
    "drp_wt = pred_ps_wt-pred_bow_wt                                    # wild-type degron regulatory potential"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Convert string to 2D list of characters\n",
    "\n",
    "seqlist = [] # 2D list of all sequences, with each character as 1 entry \n",
    "for sequence in input_seq:\n",
    "    seqlist.append(list(sequence))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define sequence information variables\n",
    "seq_length = len(seqlist[0]) # 23\n",
    "num_seq = len(seqlist)       # number of total sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Sort degron sequence information by regulatory potential in descending order (highest to lowest)\n",
    "degron_pred_sorted = degron_pred.sort_values(by=['regulatory potential'], ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. ROC curve calculation to statistically determine sequence cutoff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Determines true labels based on bins\n",
    "# Bins 1 & 2 -> 1 (high degradation)\n",
    "# Bins 3 & 4 -> 0 (low degradation)\n",
    "\n",
    "auc_info = np.asarray(degron_pred_sorted[['Modal Bin', 'sequence position specific', 'bag of words']])\n",
    "\n",
    "for i in range(num_seq):\n",
    "    mbin = auc_info[:, 0][i]\n",
    "    if mbin > 2:\n",
    "        auc_info[:, 0][i] = 0\n",
    "    else:\n",
    "        auc_info[:, 0][i] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def auc_details(cutoff):\n",
    "    y_true = auc_info[cutoff:num_seq-cutoff, 0]\n",
    "    y_score_ss = auc_info[cutoff:num_seq-cutoff, 1] # Sequence-specific prediction scores\n",
    "    y_score_bow = auc_info[cutoff:num_seq-cutoff, 2] # Bag of words prediction scores\n",
    "\n",
    "    fp_ss, tp_ss, _ = sklearn.metrics.roc_curve(y_true, y_score_ss) # false-positive and true-positive\n",
    "    auc_ss = auc(fp_ss, tp_ss)\n",
    "    fp_bow, tp_bow, _ = sklearn.metrics.roc_curve(y_true, y_score_bow)\n",
    "    auc_bow = auc(fp_bow, tp_bow) \n",
    "    \n",
    "    return [fp_ss, tp_ss, auc_ss, fp_bow, tp_bow, auc_bow]\n",
    "\n",
    "# cutoffs is a 1D list containing the cutoffs to be plotted\n",
    "def plot_auc(cutoffs):\n",
    "    colors = [ ['mediumpurple', 'indigo'], ['lightcoral', 'indianred'],\n",
    "             ['cornflowerblue', 'mediumblue'], ['orange', 'darkorange'],['green', 'darkgreen']]\n",
    "    #Accessed from: stackoverflow.com/questions/22408237/named-colors-in-matplotlib\n",
    "    \n",
    "    plt.figure()\n",
    "    for i in range(len(cutoffs)): \n",
    "        c = cutoffs[i]\n",
    "        a = auc_details(c)\n",
    "\n",
    "        lw = 1.5\n",
    "        plt.plot(a[0], a[1], color = colors[i][0],\n",
    "                 lw = lw, label = 'Cutoff ' + str(c) + ', Sequence-specific (area = %0.3f)' % a[2])\n",
    "        plt.plot(a[3], a[4], color = colors[i][1],\n",
    "                 lw = lw, label = 'Cutoff ' + str(c) + ', Bag-of-words (area = %0.3f)' % a[5])\n",
    "    \n",
    "    plt.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')\n",
    "    plt.xlim([0.0, 1.0])\n",
    "    plt.ylim([0.0, 1.05])\n",
    "    plt.xlabel('False Positive Rate')\n",
    "    plt.ylabel('True Positive Rate')\n",
    "    plt.title('Receiver operating characteristic (ROC), Cutoff = ' + str(c))\n",
    "    plt.legend(loc=\"lower right\")\n",
    "    \n",
    "    cutoffs_str = [str(i) for i in cutoffs]\n",
    "    cutoffs_file_name = '_'.join(cutoffs_str)\n",
    "    plt.savefig(wd+'figures/cutoffs/auc/'+cutoffs_file_name,dpi=300)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "elbow = 1860"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1860\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/cynthiachen/anaconda2/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:32: MatplotlibDeprecationWarning: Adding an axes using the same arguments as a previous axes currently reuses the earlier instance.  In a future version, a new instance will always be created and returned.  Meanwhile, this warning can be suppressed, and the future behavior ensured, by passing a unique label to each axes instance.\n"
     ]
    }
   ],
   "source": [
    "# Determine delta AUC values for all cutoffs to determine the minimum AUC\n",
    "\n",
    "aucs = []\n",
    "num_cutoffs = 400\n",
    "increment = 20\n",
    "cutoffs = np.asarray(range(num_cutoffs)) * increment\n",
    "\n",
    "b=False\n",
    "for c in cutoffs:\n",
    "    a = auc_details(c)\n",
    "    if (b==False and a[2] - a[5] < 0.01):\n",
    "        print(c)\n",
    "        b=True\n",
    "    aucs.append(a[2] - a[5])\n",
    "    \n",
    "fig = plt.figure()\n",
    "sns.set_style(\"darkgrid\")\n",
    "plt.plot(cutoffs, aucs)\n",
    "plt.xlabel('Cutoff')\n",
    "plt.ylabel('Delta AUC')\n",
    "\n",
    "combined = np.asarray([cutoffs, aucs])\n",
    "min_index = np.argmin(combined[1])\n",
    "seq_cutoff = int(combined[0][min_index]) # cutoff for the minimum difference betweeen SS and BOW\n",
    "\n",
    "# Points\n",
    "#plt.plot(seq_cutoff, combined[1][min_index], 'ro')\n",
    "elbow_auc = auc_details(elbow)[2]-auc_details(elbow)[5]\n",
    "plt.plot(elbow, elbow_auc, 'ro')\n",
    "\n",
    "# Add annotation\n",
    "ax = fig.add_subplot(111)\n",
    "# ax.annotate(\"Minimum \\n (Cutoff = \"+ str(seq_cutoff)+ \")\", xy=(seq_cutoff, combined[1][min_index]), \n",
    "#                   xytext=(seq_cutoff-2000, 0.025), #va=\"bottom\", ha=\"center\",\n",
    "#                   bbox=dict(boxstyle=\"round\", fc=\"w\"),\n",
    "#                   arrowprops=dict(arrowstyle=\"->\"))\n",
    "elbow = 1860\n",
    "ax.annotate(\"Elbow \\n (Cutoff = \" + str(elbow) + \")\", xy=(elbow, elbow_auc), \n",
    "                  xytext=(elbow, 0.025), #va=\"bottom\", ha=\"center\",\n",
    "                  bbox=dict(boxstyle=\"round\", fc=\"w\"),\n",
    "                  arrowprops=dict(arrowstyle=\"->\"))\n",
    "delta = u'Î”'\n",
    "plt.title(delta + ' AUC at Various Cutoffs')\n",
    "    \n",
    "plt.savefig(wd+'figures/cutoffs/auc/delta_auc_over_cutoffs',dpi=300)\n",
    "#plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.015994577802107335"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "auc_details(1000)[2]-auc_details(1000)[5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def auc_delta(i):\n",
    "    return auc_details(i)[2]-auc_details(i)[5]\n",
    "\n",
    "def secondDerivative(i):\n",
    "    return auc_delta(i+2*increment) + auc_delta(i-2*increment) - 2 * auc_delta(i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.08276618362049915\n",
      "0.32512478772400144\n",
      "0.0010881816074833628\n",
      "0.0009599087421691133\n",
      "0.0004503978503938022\n",
      "0.00014121110149467597\n",
      "-0.00028174883709219944\n",
      "0.00024471440362161534\n",
      "0.0008309260471605073\n",
      "1.3160154739288998e-05\n",
      "-0.00040318133073102835\n",
      "0.00019706849210898714\n",
      "0.0005122813332060216\n",
      "0.00029844819572244496\n",
      "0.0001472419729245944\n",
      "-5.077419731669419e-05\n",
      "-0.00037045731249041314\n",
      "2.451795748703578e-05\n",
      "0.00045373016062078\n",
      "0.00024457992967896747\n",
      "0.00023392646545650209\n",
      "0.00017272209132668248\n",
      "-0.00011324314921901113\n",
      "-7.943650616881737e-05\n",
      "-7.24176496557094e-05\n",
      "-4.435744765052263e-05\n",
      "0.0003132727518129208\n",
      "0.00029343015388350047\n",
      "-0.00011992668826354969\n",
      "-0.00014736812608195926\n",
      "8.27232286076951e-05\n",
      "0.00010714155499946809\n",
      "0.00011173111249640488\n",
      "2.5806910861736476e-05\n",
      "-8.445312900873247e-05\n",
      "-0.00019614137735424197\n",
      "-4.619313728393415e-05\n",
      "0.00027064544104427046\n",
      "0.00010280784480398708\n",
      "-0.0001079765231937202\n",
      "-0.0001512744662519605\n",
      "-9.237894135205593e-07\n",
      "0.0002558549896789497\n",
      "-4.889108149108701e-05\n",
      "-0.00019802689433545684\n",
      "0.0002603102751748132\n",
      "0.00020604901167997625\n",
      "-0.00020019458773679055\n",
      "-0.00017096044785025466\n",
      "0.00025843681451021006\n",
      "0.0003453063574374582\n",
      "0.00014936418990518785\n",
      "-7.963816442457805e-05\n",
      "-0.00022882416066349887\n",
      "5.804648904550458e-05\n",
      "0.0001193095114199183\n",
      "-6.796126708463746e-05\n",
      "-0.00020621587788161033\n",
      "-8.99189753206242e-05\n",
      "0.00011393883059374055\n",
      "-1.666135372335731e-05\n",
      "3.222255184298639e-05\n",
      "0.0001895662156617961\n",
      "-0.00013170636340531505\n",
      "-0.0003452803909322899\n",
      "9.95804833635594e-05\n",
      "0.0002445058171630654\n",
      "-6.922361172501112e-05\n",
      "-5.1005207688747944e-05\n",
      "9.898811508946004e-05\n",
      "-5.823782074476647e-06\n",
      "2.197307690354755e-06\n",
      "0.0003008555714434369\n",
      "0.00016511911064909235\n",
      "-0.0002542859112772966\n",
      "-0.0001172864142708896\n",
      "5.783296948658201e-05\n",
      "5.656235356599204e-05\n",
      "0.0002178093823009819\n",
      "0.0001680319827473964\n",
      "-7.829837290418418e-05\n",
      "-8.496025020954256e-06\n",
      "0.0002449647375325492\n",
      "-1.9552976579295667e-05\n",
      "-0.0003467807324263328\n",
      "-0.00033304173676951265\n",
      "-0.0001988182156188456\n",
      "0.00016999158409536363\n",
      "0.0002863650217498348\n",
      "6.187925866107946e-05\n",
      "-0.00015428607962209018\n",
      "-6.116503447928245e-05\n",
      "0.00017750488716650192\n",
      "-2.5605170083098372e-05\n",
      "-0.00013062606345626282\n",
      "0.00012773424559808788\n",
      "0.0001768563844765092\n",
      "-9.914793565235502e-05\n",
      "-0.00017640791010231638\n",
      "5.507970239504445e-05\n",
      "0.00013940609174156737\n",
      "7.359895854197607e-05\n",
      "-0.0001171792515040071\n",
      "-0.00010826447588940802\n",
      "0.00011082771433090777\n",
      "-3.754293282687282e-05\n",
      "-0.00012073974578974411\n",
      "0.00015677647533862693\n",
      "0.00010963277168019037\n",
      "-6.404744125032558e-05\n",
      "-4.386109097309632e-05\n",
      "-4.811182683361892e-05\n",
      "4.9843167531427035e-05\n",
      "6.826566343520746e-05\n",
      "5.451290666647068e-06\n",
      "-4.690851115729533e-05\n",
      "-3.1508964541848705e-05\n",
      "0.0001321803551008749\n",
      "0.00011837094430455775\n",
      "-2.3811159142850258e-05\n",
      "-0.00010070865163680498\n",
      "-6.117121432924755e-05\n",
      "0.00015043452241092492\n",
      "0.00010949440824947754\n",
      "-0.00013825015308854827\n",
      "-3.776074559780351e-05\n",
      "1.0746016975371475e-05\n",
      "-4.830751723772231e-05\n",
      "4.41563608304385e-05\n",
      "2.0267399201356895e-05\n",
      "2.5975086979213025e-05\n",
      "-5.977651588207067e-05\n",
      "-3.8494308178149694e-05\n",
      "0.00018142102698037732\n",
      "4.060374433345615e-06\n",
      "-5.1840825398552504e-05\n",
      "9.846999333984474e-05\n",
      "-9.443448076740246e-06\n",
      "-8.008294870531785e-05\n",
      "-6.976851499995718e-05\n",
      "-7.60998511217803e-05\n",
      "-0.00014941778931487626\n",
      "7.721913843794503e-06\n",
      "0.00015470265925643378\n",
      "-0.00014999549809702195\n",
      "-8.026643577252468e-05\n",
      "0.0002554704366701799\n",
      "0.00010457995516943708\n",
      "-5.498218130584043e-05\n",
      "7.705185956285998e-05\n",
      "6.746500953069834e-05\n",
      "-4.111292331199223e-05\n",
      "4.127286600486002e-06\n",
      "-1.4552514272070383e-05\n",
      "-6.999434067234134e-05\n",
      "-7.12140888133872e-05\n",
      "-1.0923503831805625e-05\n",
      "8.59871319641714e-05\n",
      "1.8095490235903178e-05\n",
      "4.571208769510449e-06\n",
      "3.2035812478459214e-05\n",
      "-4.2674080660609626e-05\n",
      "-9.069157380581228e-05\n",
      "-1.3108496888181165e-05\n",
      "0.00019926595995678476\n",
      "5.223386014019393e-05\n",
      "-0.00014757362186701428\n",
      "-9.161310286953661e-05\n",
      "-5.235747835907745e-05\n",
      "0.00017893949752334493\n",
      "0.00022293941502626868\n",
      "-4.874701874046927e-06\n",
      "-3.6975198152600086e-05\n",
      "-7.147055147871395e-05\n",
      "-0.00020808511921377448\n",
      "-6.587881997233147e-05\n",
      "0.00020436707680215882\n",
      "2.7054034259621496e-05\n",
      "-0.0001525296128439546\n",
      "3.744675196726277e-05\n",
      "7.321340811994315e-05\n",
      "6.241549980900984e-05\n",
      "-1.476333361216664e-05\n",
      "-0.00017064530732291416\n",
      "-6.085553212997752e-05\n",
      "1.4763842822063111e-05\n",
      "-2.687300741088805e-05\n",
      "4.2501909694903084e-05\n",
      "0.00011033586544628715\n",
      "5.427573565441879e-05\n",
      "-3.504686593713835e-05\n",
      "-2.6755264865951922e-05\n",
      "-2.9148709229653846e-06\n",
      "1.5375428941366565e-05\n",
      "1.465388404642276e-05\n",
      "-7.826000957167345e-05\n",
      "2.2298295209166596e-06\n",
      "0.0001145694270705544\n",
      "4.510738141227932e-06\n",
      "-4.273134063148554e-06\n",
      "6.665061530386485e-05\n",
      "-1.253369725240372e-05\n",
      "-8.893822097910675e-05\n",
      "2.600248292772278e-05\n",
      "0.0001006257963674928\n",
      "-0.00012432776185244965\n",
      "-0.00020784280821739554\n",
      "-0.00018255443244330483\n",
      "-6.63878197684209e-05\n",
      "0.0003151792827535971\n",
      "0.0003423018021974089\n",
      "4.7533114929998455e-05\n",
      "-0.00015045237521793542\n",
      "-0.0001522674134665225\n",
      "-3.832256052238581e-05\n",
      "0.0001095829091639633\n",
      "7.67976170156226e-05\n",
      "-0.00010884581338199695\n",
      "-3.676166678012738e-05\n",
      "-3.621093223937777e-06\n",
      "-3.998834687601871e-05\n",
      "2.726374193595671e-05\n",
      "0.0001115157226685426\n",
      "0.00012655916027815017\n",
      "8.574645100845046e-06\n",
      "2.3572518571879364e-05\n",
      "-2.477719454718752e-05\n",
      "-5.795045957235345e-05\n",
      "6.465452497161905e-06\n",
      "-0.00010182297227412374\n",
      "-8.057971527097862e-05\n",
      "2.5389839807266057e-06\n",
      "-1.7944168697825447e-05\n",
      "6.68686232043747e-05\n",
      "-4.364507037102339e-05\n",
      "-0.0001756811546245185\n",
      "0.00013549404833057554\n",
      "0.00020009936887799284\n",
      "2.830874133064043e-05\n",
      "-4.405900700021892e-05\n",
      "-3.3063318966464195e-05\n",
      "0.00011293660908495706\n",
      "-2.4317662076889235e-05\n",
      "-0.00010191333601328179\n",
      "3.860169681346726e-05\n",
      "-5.512049812306241e-05\n",
      "-0.0001672429376675666\n",
      "1.2721568152795903e-05\n",
      "0.00019090708858027217\n",
      "6.679460656144887e-05\n",
      "-0.00011225304849171724\n",
      "-4.4996868597824147e-07\n",
      "7.742573492541815e-05\n",
      "-8.868805373074551e-05\n",
      "-0.00010253909009516615\n",
      "9.948525303427047e-05\n",
      "0.00014291923061915401\n",
      "-0.00012960789877447532\n",
      "-0.00011153884501413636\n",
      "6.660812437997432e-05\n",
      "-1.764422553107714e-05\n",
      "1.346763708220955e-06\n",
      "-1.9339581473487577e-05\n",
      "-0.00011387354189151822\n",
      "4.503693697111011e-05\n",
      "0.0001372638266535775\n",
      "-2.381240722071354e-05\n",
      "-6.415573814344544e-05\n",
      "-5.1711595787096165e-05\n",
      "-8.904804066101057e-05\n",
      "0.0002016179611927793\n",
      "0.00026806853735461544\n",
      "-0.00013766609071597102\n",
      "-8.650256344799878e-05\n",
      "6.13402212633618e-05\n",
      "-1.9250833321105e-05\n",
      "-1.482453125545291e-05\n",
      "1.8835785831061713e-07\n",
      "8.124010584276142e-05\n",
      "0.00014787040845787303\n",
      "1.1231296207836472e-05\n",
      "-0.00014525416279476122\n",
      "-4.1626271628847356e-05\n",
      "-1.8542294328405262e-05\n",
      "-0.00012905196431622912\n",
      "2.4957411585702083e-05\n",
      "0.00020676559347254653\n",
      "-4.082666941762625e-05\n",
      "-0.0001557224749687558\n",
      "-1.1115905555247885e-05\n",
      "-7.00233996578925e-05\n",
      "-3.4905511173799475e-05\n",
      "0.00010501468548029802\n",
      "4.820120771020786e-05\n",
      "-0.0001312416085896828\n",
      "-2.27657858731245e-06\n",
      "0.00021437440230587246\n",
      "0.00012996681815147237\n",
      "-6.97796656872196e-05\n",
      "-5.553926298484946e-05\n",
      "0.00013107992226035758\n",
      "-4.2521924108362974e-05\n",
      "-0.0002466073498982313\n",
      "-2.2583141149112862e-05\n",
      "0.000154343396770118\n",
      "-5.378199974259701e-06\n",
      "-9.21425331023018e-05\n",
      "4.155857429433141e-05\n",
      "-2.983660023747614e-05\n",
      "-0.00011229975994975216\n",
      "0.00012691692882871308\n",
      "0.00020914606769462107\n",
      "-4.40239632075734e-05\n",
      "-4.962327319391058e-05\n",
      "6.160509056596375e-05\n",
      "-9.357562261969044e-05\n",
      "8.781549071912309e-06\n",
      "4.4471048305072536e-05\n",
      "-0.00022551379011503947\n",
      "-9.787663422367032e-05\n",
      "0.00017385343413667265\n",
      "0.00023103654355127734\n",
      "0.00010834839041340416\n",
      "-2.6102760456225838e-05\n",
      "-6.880107370332844e-05\n",
      "-0.0001341786868599426\n",
      "-6.092309501226634e-05\n",
      "2.0568389381692143e-05\n",
      "-0.00010123123132677048\n",
      "1.2379680402130333e-05\n",
      "0.000296333593864162\n",
      "5.255866632902784e-05\n",
      "-0.00015126449278279885\n",
      "5.841452723609475e-05\n",
      "9.548940076409274e-05\n",
      "-8.803268822477062e-05\n",
      "-0.0002560223740017964\n",
      "-0.00014602235685556852\n",
      "0.0001437637779601042\n",
      "0.00019097705189463632\n",
      "-6.768222083475273e-05\n",
      "-2.1130256262713587e-05\n",
      "0.00013944585465774928\n",
      "2.92911071692048e-05\n",
      "9.932522955291567e-05\n",
      "-0.0009132629080229604\n",
      "-0.0011594791019275963\n",
      "0.0007821544121053847\n",
      "0.0007059338727909426\n",
      "-0.0001679183681783103\n",
      "0.00020634430077692834\n",
      "6.969642718834823e-05\n",
      "7.056028540730619e-05\n",
      "9.075469575314798e-05\n",
      "-0.00014468960038460654\n",
      "7.643810775481708e-05\n",
      "0.0003021029487207949\n",
      "9.835518126577458e-05\n",
      "-0.0001526998473193597\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.00010010302655694137\n",
      "0.0004959594965324721\n",
      "0.00024815196111893645\n",
      "-0.0002785029967264929\n",
      "-0.00031311060568428584\n",
      "-0.0002569163340380154\n",
      "-0.0006402866239750837\n",
      "-0.00043160524988938853\n",
      "0.0006353726497910772\n",
      "0.0005873948378869454\n",
      "-0.00015081596262789265\n",
      "-8.05809261276158e-05\n",
      "5.6051288130798405e-05\n",
      "-0.00012832723163802573\n",
      "-0.00041662909159312367\n",
      "-0.00043196822971647286\n",
      "0.0001300871417889704\n",
      "0.0006377856096256851\n",
      "0.00041321777465885834\n",
      "-0.00013430364689515084\n",
      "-0.00024005263739190763\n",
      "-2.3545366747423913e-05\n",
      "4.245527467383603e-05\n",
      "-7.996834795442886e-05\n",
      "-0.001482707683602591\n",
      "-0.0014803587347982372\n",
      "0.001179785964223301\n",
      "0.0013923620232019207\n",
      "9.205939348655434e-05\n",
      "-1.065234286157235e-05\n",
      "0.0009630129428862544\n",
      "0.0009038563515381126\n",
      "-0.0006481862835194807\n",
      "-0.0007644912198901954\n",
      "-0.00020722938243056266\n",
      "0.0002056222889621484\n",
      "0.0001474632734907999\n",
      "-0.00017913870173391633\n",
      "1.4335964876543983e-05\n",
      "-9.659476838519421e-05\n",
      "-0.0002184087765044751\n"
     ]
    }
   ],
   "source": [
    "min_deriv = 100\n",
    "for c in cutoffs:\n",
    "    print(secondDerivative(c))\n",
    "    if (secondDerivative(c) < min_deriv):\n",
    "        knee_point = c\n",
    "        min_deriv = secondDerivative(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.001482707683602591"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_deriv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# x = range(1,21)\n",
    "# y = [0.065, 0.039, 0.030, 0.024, 0.023, 0.022, 0.019, 0.0185, 0.0187,\n",
    "#      0.016, 0.015, 0.016, 0.0135, 0.0130, 0.0125, 0.0120, 0.0117, 0.0115, 0.0112, 0.013]\n",
    "\n",
    "# kn = KneeLocator(\n",
    "#     x,\n",
    "#     y,\n",
    "#     curve='convex',\n",
    "#     direction='decreasing',\n",
    "#     interp_method='interp1d',\n",
    "# )\n",
    "\n",
    "# print(kn.knee)\n",
    "# 7\n",
    "# import matplotlib.pyplot as plt\n",
    "# plt.xlabel('x')\n",
    "# plt.ylabel('f(x)')\n",
    "# plt.xticks(range(1,21))\n",
    "# plt.plot(x, y, 'bx-')\n",
    "# plt.vlines(kn.knee, plt.ylim()[0], plt.ylim()[1], linestyles='dashed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seq_cutoff =1860"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Plot ROC curve for cutoff with minimum delta AUC in comparison to original AUC\n",
    "\n",
    "plot_auc([0, int(seq_cutoff/2), seq_cutoff])\n",
    "plot_auc([0, 2000])\n",
    "plot_auc([0, seq_cutoff])\n",
    "plot_auc([seq_cutoff])\n",
    "plot_auc([0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## 2. Motif Analysis\n",
    "Discover patterns that characterize sequences with high & low degron regulatory potential"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Select only the \"sequence\" and \"potential\" columns and format values into a list\n",
    "degron_seq_scores = degron_pred_sorted[['Peptide amino acid sequence','regulatory potential']].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-5'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(-5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# MOTIF ANALYSIS FUNCTIONS\n",
    "\n",
    "\n",
    "# Function description: Determine the frequency of each kmer within the dataset\n",
    "#\n",
    "# Cutoff -- number of sequences from the top + bottom to consider for motif analysis\n",
    "# kmer length -- length of kmers to be considered\n",
    "# terminal -- how many amino acids to consider for each sequence (starting from the end)\n",
    "#\n",
    "# Returns a dictionary (key: kmer, value: frequency)\n",
    "\n",
    "def kmer_freq(cutoff, kmer_length, terminal = seq_length):\n",
    "    top = degron_seq_scores[0:cutoff]\n",
    "    bottom = degron_seq_scores[-cutoff:]\n",
    "    kmers = []\n",
    "    kmer_dict = {}\n",
    "    \n",
    "    for row in top:\n",
    "        sequence = row[0]\n",
    "        for j in range(seq_length - terminal, seq_length - kmer_length + 1):\n",
    "            kmers.append((sequence[j:(j + kmer_length)]+'('+str(j-seq_length)+')', row[1])) \n",
    "            # add k-mer and corresponding score\n",
    "    \n",
    "    # Normalize drp scores to emphasize differences\n",
    "    kmers = np.asarray(kmers) # Convert kmers to np array from a list\n",
    "    x = np.asarray(list(map(float, kmers[:, 1])))\n",
    "    normalized = (x-min(x))/(max(x)-min(x))\n",
    "    kmers[:,1] = normalized\n",
    "    \n",
    "    # Convert list of kmers into 1 string\n",
    "    kmer_str = \" \".join(kmers[:, 0])\n",
    "    kmer_list = list(kmers[:, 0])\n",
    "    \n",
    "    unique_kmers = [] \n",
    "  \n",
    "    # Determine all unique kmers\n",
    "    for i in kmer_list: \n",
    "        # checking for duplicacy \n",
    "        if i not in unique_kmers: \n",
    "            # insert value in str2 \n",
    "            unique_kmers.append(i)  \n",
    "    \n",
    "    for i in range(0, len(unique_kmers)): \n",
    "        # count the frequency of each unique kmer and print \n",
    "        #print('Frequency of', unique_kmers[i], 'is :', kmer_list.count(unique_kmers[i]))\n",
    "        kmer_dict.update({unique_kmers[i] : kmer_list.count(unique_kmers[i])})\n",
    "    \n",
    "    # Returns dictionary of [kmer sequence] -> [frequency]\n",
    "    return kmer_dict\n",
    "\n",
    "\n",
    "# Function description: Given a dataset of kmer information and a single specified kmer, \n",
    "# this function determines binomial probability values for various cutoffs \n",
    "# to determine the optimal motif cutoff frequency\n",
    "# ???\n",
    "def graph_binom(kmer, kmer_info):\n",
    "    plt.close()\n",
    "    # Retrieve kmer information from compiled dictionary\n",
    "    info = kmer_info[kmer]\n",
    "    top_count = info[2]\n",
    "    baseline_prob = info[1] \n",
    "    p_value = info[3]\n",
    "    \n",
    "    # Determine frequency range to be plotted\n",
    "    #middle = int(baseline_prob * seq_cutoff)\n",
    "    middle = top_count\n",
    "    freqs = range(middle - 20, middle + 20, 1)\n",
    "    \n",
    "    # Plot kmer-specific binomial curve\n",
    "    binom_vals = []\n",
    "    for i in freqs:\n",
    "        binom_vals.append(stats.binom.sf(i, seq_cutoff, baseline_prob))\n",
    "    \n",
    "    fig = plt.figure()\n",
    "    plt.plot(freqs, binom_vals, 'b', label = \"Binomial expression\")\n",
    "\n",
    "    # Add annotations\n",
    "    \n",
    "    # horizontal ray\n",
    "    plt.axhline(y=p_value, xmax=0.5, color='r', linestyle='--', label = \"y = p-value\", lw = 1)\n",
    "    # vertical ray\n",
    "    #plt.axvline(x=top_count, ymax=p_value, color='r', linestyle='--', label = \"x = count\", lw = 1)\n",
    "    #plt.plot(pair_x, pair_y, 'm', linewidth=3)\n",
    "\n",
    "    # based on top_freq\n",
    "    ax = fig.add_subplot(111)\n",
    "\n",
    "    plt.plot(top_count, p_value, 'ro') # point\n",
    "    \n",
    "    # plt.plot(motif_cutoff, binom_value, 'ro') #point\n",
    "    # ax.annotate(\"Optimal Motif Frequency \\n Cutoff = \" + str(motif_cutoff), \n",
    "    # xy=(motif_cutoff, binom_value), \n",
    "    #                   xytext=(motif_cutoff-8, 0.4), #va=\"bottom\", ha=\"center\",\n",
    "    #                   bbox=dict(boxstyle=\"round\", fc=\"w\"),\n",
    "    #                   arrowprops=dict(arrowstyle=\"->\"))\n",
    "\n",
    "    plt.xlabel('Frequency Count')\n",
    "    plt.ylabel('Binomial Survival Expression')\n",
    "    plt.legend(loc=\"upper right\")\n",
    "    plt.title(\"Motif-Specific Binomial Survival Expression over Frequency Cutoffs: \\n Motif = \" + kmer)\n",
    "    plt.savefig(wd+'figures/cutoffs/binomial/'+kmer,dpi=300)\n",
    "    plt.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# gap is the position where the gap occurs\n",
    "# Example: if AAA_A, then gap = 4\n",
    "def kmer_freq_with_gap(cutoff, gap, kmer_length, terminal = seq_length):\n",
    "    top = degron_seq_scores[0:cutoff]\n",
    "    bottom = degron_seq_scores[-cutoff:]\n",
    "    kmers = []\n",
    "    kmer_dict = {}\n",
    "    \n",
    "    for row in top:\n",
    "        sequence = row[0]\n",
    "        for j in range(seq_length - terminal, seq_length - kmer_length + 1):\n",
    "            kmer_str = sequence[j:(j+gap-1)] + '_' + sequence[(j+gap):(j+kmer_length)]\n",
    "            kmers.append((kmer_str + str(j+1), row[1])) # add k-mer and corresponding score\n",
    "    \n",
    "    # Normalize drp scores to emphasize differences\n",
    "    kmers = np.asarray(kmers) # Convert kmers to np array from a list\n",
    "    x = np.asarray(list(map(float, kmers[:, 1])))\n",
    "    normalized = (x-min(x))/(max(x)-min(x))\n",
    "    kmers[:,1] = normalized\n",
    "    \n",
    "    # Convert list of kmers into 1 string\n",
    "    kmer_str = \" \".join(kmers[:, 0])\n",
    "    kmer_list = list(kmers[:, 0])\n",
    "    \n",
    "    unique_kmers = [] \n",
    "  \n",
    "    # Determine all unique kmers\n",
    "    for i in kmer_list: \n",
    "        # checking for duplicacy \n",
    "        if i not in unique_kmers: \n",
    "            # insert value in str2 \n",
    "            unique_kmers.append(i)  \n",
    "    \n",
    "    for i in range(0, len(unique_kmers)): \n",
    "        # count the frequency of each unique kmer and print \n",
    "        #print('Frequency of', unique_kmers[i], 'is :', kmer_list.count(unique_kmers[i]))\n",
    "        kmer_dict.update({unique_kmers[i] : kmer_list.count(unique_kmers[i])})\n",
    "    \n",
    "    # Returns dictionary of [kmer sequence] -> [frequency]\n",
    "    return kmer_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. K-mer probability calculation to statistically determine significant motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "seq_cutoff = 1860\n",
    "kmer_length = 1\n",
    "t = 6\n",
    "\n",
    "# kmer_dict = kmer_freq_with_gap(cutoff = num_seq, gap=2, kmer_length = kmer_length, terminal = t) # not excluding any sequences\n",
    "# kmer_dict_top = kmer_freq_with_gap(cutoff = seq_cutoff, gap=2, kmer_length = kmer_length, terminal = t)\n",
    "kmer_dict = kmer_freq(cutoff = num_seq, kmer_length = kmer_length, terminal = t) # not excluding any sequences\n",
    "kmer_dict_top = kmer_freq(cutoff = seq_cutoff, kmer_length = kmer_length, terminal = t)\n",
    "\n",
    "# Create motif information dictionary:\n",
    "# Key: kmer\n",
    "# Values (kmer-specific):\n",
    "# -- 1) total count (frequency of kmer among all sequences)\n",
    "# -- 2) baseline probability, \n",
    "# -- 3) top count (frequency of kmer among \"top\" sequences as determined by seq_cutoff, \n",
    "# -- 4) p-value\n",
    "\n",
    "kmer_info = {}\n",
    "\n",
    "for kmer in kmer_dict_top:\n",
    "    total_count = kmer_dict[kmer]\n",
    "    baseline_prob = total_count / num_seq\n",
    "    top_count = kmer_dict_top[kmer] # Frequency kmer among top sequences \n",
    "    p_value = stats.binom.sf(top_count - 1, seq_cutoff, baseline_prob) # Subtract 1 because of inclusitivity\n",
    "    kmer_info.update({kmer : [total_count, baseline_prob, top_count, p_value]})              \n",
    "                      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#graph_binom('EED18', kmer_info = kmer_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Benjamini Hochberg p-value adjustment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Compile list of all p-values\n",
    "p_values = []\n",
    "for i in kmer_info:\n",
    "    p_values.append(str(kmer_info[i][3]))\n",
    "    \n",
    "# Convert list of p-values to csv for R processing\n",
    "pd.DataFrame(p_values).transpose().to_csv('p_values.csv', header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#%run bh.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# AT THIS TIME: RUN bh.ipynb FILE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read in bh values from file\n",
    "bh_p_values = pd.read_csv('bh_p_values.csv', header=None)[0]\n",
    "\n",
    "# Add bh pvalues to kmer info table\n",
    "counter = 0\n",
    "for i in kmer_info:\n",
    "    kmer_info[i].append(bh_p_values[counter])\n",
    "    counter=counter+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determining top significant motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "kmer_info_sorted = sorted(kmer_info.items(), key=lambda item: item[1][4])\n",
    "\n",
    "i = 0\n",
    "# Use BH p-values to determine significant motifs\n",
    "while (kmer_info_sorted[i][1][4] <= 0.05):\n",
    "    i=i+1\n",
    "motif_cutoff = i\n",
    "#length_significance.append(motif_cutoff) # for analysis on kmer length vs. # sig. motifs\n",
    "    \n",
    "top_motifs = kmer_info_sorted[0:motif_cutoff]\n",
    "top_sequences = degron_seq_scores[0:seq_cutoff]\n",
    "\n",
    "for i in range(motif_cutoff):\n",
    "    top_motifs[i] = [top_motifs[i][0]]+top_motifs[i][1]\n",
    "    \n",
    "top_motifs = pd.DataFrame(top_motifs, columns = ['Motif', 'total count', 'baseline probability', 'top count', 'p-value', 'BH p-value'])\n",
    "\n",
    "filename = 'motif_info/'+str(kmer_length)+'mers.csv'\n",
    "pd.DataFrame(kmer_info_sorted).to_csv(filename, header=False, index=False)\n",
    "filename = 'motif_info/'+str(kmer_length)+'mers_top.csv'\n",
    "pd.DataFrame(top_motifs).to_csv(filename, header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Plot Number of significant motifs vs. Motif length\n",
    "# length_significance = pd.DataFrame([[2, 94], [3, 5], [4, 2], [5, 0], [6, 0]], \n",
    "#                                    columns = ['Motif Length', 'Number of Significant Motifs'])\n",
    "\n",
    "# ax = sns.swarmplot(data = length_significance, x ='Motif Length', y = 'Number of Significant Motifs' )\n",
    "# plt.title('Number of Significant Motifs vs. Motif Length')\n",
    "# #plt.show()\n",
    "\n",
    "# plt.savefig(wd+'figures/motifs/length_vs_significance',dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Sequence logo visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Identification of all sequences that contain motifs\n",
    "sequences = []\n",
    "for identifier in list(top_motifs[\"Motif\"]):\n",
    "    motif = identifier[0:kmer_length]\n",
    "    index = seq_length + int(identifier[kmer_length+1:-1])\n",
    "    contains_motif = []\n",
    "    for i in range(seq_cutoff):\n",
    "        if degron_seq_scores[i][0].find(motif) == index:\n",
    "            contains_motif.append(degron_seq_scores[i][0])\n",
    "    sequences.append(contains_motif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'G(-1)'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_motifs['Motif'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G(-1)\n",
      "KFSQRG\n",
      "QALRRG\n",
      "SLESRG\n",
      "RNATRG\n",
      "HSWVSG\n",
      "NSEAVG\n",
      "VICERG\n",
      "QSLKAG\n",
      "SKALCG\n",
      "RTCRKG\n",
      "KVAVKG\n",
      "RFLPSG\n",
      "PVSPTG\n",
      "PCPSVG\n",
      "ILTKHG\n",
      "PYFPSG\n",
      "PFILEG\n",
      "LKFRFG\n",
      "PYLSSG\n",
      "AKRAWG\n",
      "HMQSYG\n",
      "AMQIDG\n",
      "MAQPVG\n",
      "RPRPWG\n",
      "SASDSG\n",
      "LMSSLG\n",
      "FRLRLG\n",
      "KEDEEG\n",
      "LKKSRG\n",
      "TTISRG\n",
      "YKVDWG\n",
      "FSPCYG\n",
      "EHPYFG\n",
      "KVKPVG\n",
      "QSCQEG\n",
      "PYYPSG\n",
      "ISLSPG\n",
      "DVRITG\n",
      "YIHTLG\n",
      "SSYAAG\n",
      "LAQFPG\n",
      "VQTWKG\n",
      "SACISG\n",
      "ILASVG\n",
      "WKEVDG\n",
      "LRVRRG\n",
      "IVSSHG\n",
      "LLEQEG\n",
      "EDPYPG\n",
      "LERLYG\n",
      "\n",
      "\n",
      "V(-2)\n",
      "NEAAVW\n",
      "TDTRVV\n",
      "HIPLVT\n",
      "LPQEVF\n",
      "AEIRVK\n",
      "LAWSVS\n",
      "RQLPVS\n",
      "GPLGVR\n",
      "REARVQ\n",
      "RGTSVF\n",
      "EALLVC\n",
      "QEPLVQ\n",
      "PESAVA\n",
      "TICSVS\n",
      "RSSLVG\n",
      "HKCEVW\n",
      "DLGSVC\n",
      "RIHNVT\n",
      "GNKPVC\n",
      "PDQKVC\n",
      "DTFKVS\n",
      "KTNDVR\n",
      "KGCEVS\n",
      "CDSPVF\n",
      "KKCRVW\n",
      "ENKPVF\n",
      "KGCEVS\n",
      "ELYAVT\n",
      "SESAVS\n",
      "YAPQVR\n",
      "PYQYVV\n",
      "PENSVG\n",
      "PQEFVV\n",
      "PRLPVF\n",
      "TRCCVM\n",
      "NEGYVY\n",
      "GNKAVR\n",
      "SMTEVT\n",
      "AAEHVK\n",
      "LSTHVS\n",
      "MAQPVG\n",
      "TETSVI\n",
      "EGNTVV\n",
      "GSSPVK\n",
      "RESLVV\n",
      "PKQGVH\n",
      "STTFVR\n",
      "ELSKVN\n",
      "PPTTVH\n",
      "YSIKVH\n",
      "ELQKVR\n",
      "PMTNVS\n",
      "QIIRVS\n",
      "AISSVK\n",
      "PRNPVH\n",
      "NMESVL\n",
      "ASWGVS\n",
      "PFEFVS\n",
      "SSPAVA\n",
      "GGINVV\n",
      "LCLLVY\n",
      "ENETVS\n",
      "NLKHVV\n",
      "CGCSVA\n",
      "METFVS\n",
      "RPTLVA\n",
      "HSETVV\n",
      "YHIPVF\n",
      "ELQKVR\n",
      "ERCRVW\n",
      "LPIQVI\n",
      "ALDQVN\n",
      "\n",
      "\n",
      "A(-1)\n",
      "GSVRTA\n",
      "GTVSGA\n",
      "SLGGGA\n",
      "ELIFYA\n",
      "ELIFYA\n",
      "HTTGKA\n",
      "RPQVLA\n",
      "IVEVPA\n",
      "LLPTKA\n",
      "VSPVKA\n",
      "KPLRTA\n",
      "VELEQA\n",
      "MMVDSA\n",
      "IHSFTA\n",
      "SDLLSA\n",
      "CLLLQA\n",
      "PLDFIA\n",
      "PLPPRA\n",
      "VQQVLA\n",
      "QSFISA\n",
      "NMQVYA\n",
      "TQPFKA\n",
      "LPGGSA\n",
      "EFEFEA\n",
      "FLEKYA\n",
      "KKLSKA\n",
      "LLFISA\n",
      "RYTEEA\n",
      "SEFVVA\n",
      "SKQHKA\n",
      "VSKRLA\n",
      "MYGDMA\n",
      "SILTSA\n",
      "IQKLTA\n",
      "IWVGDA\n",
      "GFGTQA\n",
      "EMLKQA\n",
      "EIQILA\n",
      "RPTLVA\n",
      "QVYLPA\n",
      "SIPRLA\n",
      "\n",
      "\n",
      "R(-3)\n",
      "VEIREE\n",
      "LPNRRH\n",
      "GLTREE\n",
      "QALRRG\n",
      "GSVRTA\n",
      "VLNRCV\n",
      "QDVRQS\n",
      "KASRNV\n",
      "IYNRAR\n",
      "KKKRVY\n",
      "EMGRRE\n",
      "PTGRLA\n",
      "NELRLC\n",
      "FHHRCN\n",
      "KAIRSK\n",
      "PGQRRE\n",
      "GCCRIS\n",
      "LEVRSW\n",
      "PMFRIV\n",
      "SMLRCE\n",
      "FKMREV\n",
      "GIDREF\n",
      "IVQRLT\n",
      "LPDRQR\n",
      "CVYRLQ\n",
      "STYRVS\n",
      "EVYRGN\n",
      "TINRSC\n",
      "CLPRGC\n",
      "EKNRGS\n",
      "LVGRAS\n",
      "DPKRLM\n",
      "IYLRGK\n",
      "SEVRTQ\n",
      "ITYRGT\n",
      "SKSRRS\n",
      "SKSRRS\n",
      "IQVRPL\n",
      "MKVRPA\n",
      "QIIRVS\n",
      "HGIRAW\n",
      "VLPRLL\n",
      "CPGREP\n",
      "KLERLH\n",
      "KNKRFP\n",
      "EGIRNI\n",
      "LETRQE\n",
      "LFGRSY\n",
      "\n",
      "\n",
      "A(-2)\n",
      "YQVPAY\n",
      "TEVGAF\n",
      "FEDFAC\n",
      "LEPGAA\n",
      "QSQTAY\n",
      "SRTGAS\n",
      "QPVMAN\n",
      "IYNRAR\n",
      "MSQHAQ\n",
      "HPSLAS\n",
      "VLITAF\n",
      "PTIKAG\n",
      "EPVKAK\n",
      "LEEIAC\n",
      "QSNVAS\n",
      "KLMTAL\n",
      "VLHIAT\n",
      "GNLTAF\n",
      "HQVPAY\n",
      "ETRVAD\n",
      "QPLLAT\n",
      "RIFTAD\n",
      "RGMDAV\n",
      "GDKGAG\n",
      "DDEIAV\n",
      "STPEAS\n",
      "KLQLAL\n",
      "GLVPAC\n",
      "QTINAK\n",
      "LQNHAF\n",
      "CRCQAR\n",
      "PQPGAC\n",
      "PFQMAT\n",
      "VTEIAG\n",
      "CGPLAN\n",
      "PNSVAI\n",
      "PPGEAA\n",
      "ITRPAK\n",
      "GCCVAH\n",
      "KGETAV\n",
      "LKSTAF\n",
      "PQNSAF\n",
      "SLDSAD\n",
      "GRNQAV\n",
      "VVFDAR\n",
      "TMDCAT\n",
      "DLYLAV\n",
      "PGPQAS\n",
      "QPGLAT\n",
      "IRLNAV\n",
      "WGPLAV\n",
      "PVRPAC\n",
      "SSPLAT\n",
      "GGGGAN\n",
      "HYYTAF\n",
      "RPSTAH\n",
      "QPPPAC\n",
      "VIMPAN\n",
      "TGHRAL\n",
      "RFPYAA\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in range(5):\n",
    "    print(top_motifs['Motif'][i])\n",
    "    for j in sequences[i]:\n",
    "        print(j[-6:])\n",
    "    print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. In-silico mutagenesis of motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# MUTATION FUNCTIONS\n",
    "\n",
    "# input_seq is a list of sequences to be inputted into deepDegron\n",
    "# Returns the respective DRP values for each of the sequences\n",
    "def get_drp(input_seq):\n",
    "    input_seq = pd.Series(input_seq)\n",
    "\n",
    "    # Use compute_feature_matrix function to encode sequences\n",
    "    features = compute_feature_matrix(input_seq, 6, True)\n",
    "    condensed_features = features[:, 0:20] \n",
    "    # Remove zeros at the end for bag of words prediction\n",
    "\n",
    "    # Use pre-trained model to predict sequence degredataion\n",
    "    pred_bow = model_bag_of_words.predict_proba(condensed_features) \n",
    "    pred_ps = model_pos_specific.predict_proba(features)\n",
    "    drp = pred_ps-pred_bow\n",
    "    return drp\n",
    "\n",
    "# Heatmap visualization\n",
    "def plot_weights(weights, plot_title, motif):\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))\n",
    "    im = ax.imshow(weights, cmap='viridis')\n",
    "    k = len(motif)\n",
    "\n",
    "    # Set tick labels\n",
    "    residues = list(motif)\n",
    "    ax.set_xticks(np.arange(k))\n",
    "    ax.set_yticks(np.arange(len(aminoacids)))\n",
    "    ax.set_xticklabels(residues)\n",
    "    ax.set_yticklabels(aminoacids);\n",
    "\n",
    "    # Rotate the tick labels and set their alignment.\n",
    "    # plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\", rotation_mode=\"anchor\")\n",
    "    cbar = plt.colorbar(im)\n",
    "\n",
    "    # Turn spines off and create white grid\n",
    "    for edge, spine in ax.spines.items():\n",
    "        spine.set_visible(False)\n",
    "    ax.set_xticks(np.arange(weights.shape[1]+1)-.5, minor=True)\n",
    "    ax.set_yticks(np.arange(weights.shape[0]+1)-.5, minor=True)\n",
    "    ax.grid(which=\"minor\", color=\"w\", linestyle='-', linewidth=3)\n",
    "    ax.tick_params(which=\"minor\", bottom=False, left=False)\n",
    "\n",
    "    # for tick in ax.xaxis.get_major_ticks():\n",
    "    #     tick.label.set_fontsize(14) \n",
    "    # for tick in ax.yaxis.get_major_ticks():\n",
    "    #     tick.label.set_fontsize(14) \n",
    "\n",
    "    # set size of tick labels\n",
    "    plt.tick_params(axis='both', which='major', labelsize=18)\n",
    "    cbar.ax.tick_params(labelsize=18) \n",
    "    \n",
    "    # set title\n",
    "    plt.title(plot_title, fontsize=20)\n",
    "    \n",
    "    from matplotlib import rcParams\n",
    "    rcParams['axes.titlepad'] = 20 \n",
    "\n",
    "#     plt.show()\n",
    "    plt.savefig(wd+'figures/mutations/'+motif,dpi=300)\n",
    "    plt.close()\n",
    "    \n",
    "# Given a sequence and the starting mutation index, this function returns an array\n",
    "# that contains the DRP differences\n",
    "# The array size is 20 x K, where K is the length of the motif.\n",
    "def mutate(sequence, index):\n",
    "    # Mutate sequences and calculate new DRP values\n",
    "    mutated_drp = []\n",
    "    for i in range(kmer_length):\n",
    "        position = index + i\n",
    "        mutated_seq = []\n",
    "        for letter in aminoacids:\n",
    "            mutated_seq.append(sequence[:position] + letter + sequence[(position+1):])\n",
    "        mutated_drp.append(get_drp(mutated_seq)[:, 0])\n",
    "\n",
    "    # Calculate DRP differences after mutating\n",
    "    original_drp = get_drp(sequence)\n",
    "    mutated_drp = mutated_drp - np.full((len(mutated_drp), len(aminoacids)), original_drp)\n",
    "    return mutated_drp\n",
    "\n",
    "# index1 and index2 are the two positions to mutate\n",
    "# index1 < index2\n",
    "def mutate_double(sequence, index1, index2):\n",
    "    # Mutate sequences and calculate new DRP values\n",
    "    mutated_drp = []\n",
    "    for letter1 in aminoacids:\n",
    "        mutated_seq = []\n",
    "        for letter2 in aminoacids:\n",
    "            newseq = sequence[:index1] + letter1 + sequence[(index1+1):index2] + letter2 + sequence[(index2+1):]\n",
    "            mutated_seq.append(newseq)\n",
    "        mutated_drp.append(get_drp(mutated_seq)[:, 0])\n",
    "\n",
    "    # Calculate DRP differences after mutating\n",
    "    original_drp = get_drp(sequence)\n",
    "    mutated_drp = mutated_drp - np.full((len(mutated_drp), len(aminoacids)), original_drp)\n",
    "    return mutated_drp\n",
    "\n",
    "# index1, index2, index3 are the 3 positions to mutate\n",
    "# index1 < index2 < index3\n",
    "def mutate_triple(sequence, index1, index2, index3):\n",
    "    # Mutate sequences and calculate new DRP values\n",
    "    mutated_drp1 = []\n",
    "    for letter1 in aminoacids:\n",
    "        mutated_drp2 = []\n",
    "        for letter2 in aminoacids:\n",
    "            mutated_seq = []\n",
    "            for letter3 in aminoacids:\n",
    "                newseq = sequence[:index1] + letter1 + sequence[(index1+1):index2] + letter2 + sequence[(index2+1):index3] + letter3 + sequence[(index3+1):]     \n",
    "                mutated_seq.append(newseq)\n",
    "            mutated_drp2.append(get_drp(mutated_seq)[:, 0])\n",
    "        mutated_drp1.append(mutated_drp2)\n",
    "\n",
    "    # Calculate DRP differences after mutating\n",
    "    original_drp = get_drp(sequence)\n",
    "    mutated_drp = mutated_drp1 - np.full((len(aminoacids), len(aminoacids), len(aminoacids)), original_drp)\n",
    "    return mutated_drp\n",
    "\n",
    "# Create amino acid alphabet\n",
    "aa_list='ACDEFGHIKLMNPQRSTVWY'\n",
    "aminoacids = []\n",
    "for i in range(20):\n",
    "    aminoacids.append(str(aa_list[i]))\n",
    "    \n",
    "def plot_weights_double(weights, plot_title, sequence, index1, index2):\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))\n",
    "    im = ax.imshow(weights, cmap='viridis')\n",
    "    k = len(motif)\n",
    "\n",
    "    # Set tick labels and axis labels\n",
    "    ax.set_xticks(np.arange(len(aminoacids)))\n",
    "    ax.set_yticks(np.arange(len(aminoacids)))\n",
    "    ax.set_xticklabels(aminoacids)\n",
    "    ax.set_yticklabels(aminoacids)\n",
    "    ax.set_xlabel('Alternate AA (' + sequence[index1] + ' Position ' +str(index1) + ')')\n",
    "    ax.set_ylabel('Alternate AA (' + sequence[index2] + ' Position ' +str(index2) + ')')\n",
    "\n",
    "    # plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\", rotation_mode=\"anchor\")\n",
    "    cbar = plt.colorbar(im) \n",
    "    \n",
    "    # Turn spines off and create white grid\n",
    "    for edge, spine in ax.spines.items():\n",
    "        spine.set_visible(False)\n",
    "    ax.set_xticks(np.arange(weights.shape[1]+1)-.5, minor=True)\n",
    "    ax.set_yticks(np.arange(weights.shape[0]+1)-.5, minor=True)\n",
    "    ax.grid(which=\"minor\", color=\"w\", linestyle='-', linewidth=3)\n",
    "    ax.tick_params(which=\"minor\", bottom=False, left=False)\n",
    "\n",
    "    # set size of tick labels\n",
    "    plt.tick_params(axis='both', which='major', labelsize=18)\n",
    "    cbar.ax.tick_params(labelsize=18) \n",
    "    \n",
    "    # set title\n",
    "    plt.title(plot_title, fontsize=20)\n",
    "    \n",
    "    from matplotlib import rcParams\n",
    "    rcParams['axes.titlepad'] = 20 \n",
    "\n",
    "#     plt.show()\n",
    "    plt.savefig(wd+'figures/mutations/'+motif+'_double_'+str(index1)+'_'+str(index2),dpi=300)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analysis on a single top sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Determine motif and position index from the combined info\n",
    "identifier = top_motifs['Motif'][0]\n",
    "motif = identifier[0:kmer_length]\n",
    "index = seq_length + int(identifier[kmer_length+1:-1])\n",
    "\n",
    "# Automated identification of sequences with highest DRP that contains the motif\n",
    "for i in range(len(degron_seq_scores)):\n",
    "    if degron_seq_scores[i][0].find(motif) == index:\n",
    "        sequence = degron_seq_scores[i][0]\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ELENLAAMDLELQKIAEKFSQRG'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Compute mutated DRP matrix\n",
    "mutated_drp = mutate(sequence = sequence, index = index)\n",
    "\n",
    "# Plot DRP differences using a heatmap and save the figure\n",
    "plot_weights(weights = np.transpose(np.asarray(mutated_drp)),\n",
    "             plot_title = motif+str(index+1), motif = motif)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analysis on multiple sequences that contain the motif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXUAAAElCAYAAAAbc3I/AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0\ndHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3de0AU5f4G8GeW+11RERRRRAWDTEND\nDTx5FwQR08oS65dpF0M71jHLStNIj5lHzaOZp7TM1LynaeUtzTvhDVAwAeWO3O+33X1/f5ArKwir\nrKwOz+e/mZ2Z/b6z8OzsOzPvSEIIASIikgWFoQsgIiL9YagTEckIQ52ISEYY6kREMsJQJyKSEYY6\nEZGMGBu6ANK/lJQUDB06FN26dQMAqNVqWFlZYeLEiQgICAAAbN++HeHh4XB2dgYACCFQXFyM3r17\nY/78+TAzM0NoaChSU1NhY2MDSZJQVVUFLy8vzJ07FxYWFlrv+cUXX2DDhg1o27YtAECpVMLd3R3T\npk1Dp06dAADBwcFYv349rKys8OabbyIhIQGhoaGwtLTE8uXL4ebmhq+//rqJ9tKDT5fP6G5lZmZi\n+vTp2LRpU63X5s2bh5YtWyIsLKzRtZMBCZKd5ORk0bNnT615KSkpYsiQIeKXX34RQgixbds2MWXK\nFK1lysvLRUhIiNi4caMQQogJEyaIffv2aV5Xq9UiLCxMLFy4sNZ7Ll++XHz88cda83bs2CEGDBgg\nioqKtOanpqYKLy8voVQqhRBChIaGip07d95ja+VLl89Inz7++GOxfPlyvW+Xmha7X5qJ9u3bY9q0\nafUeCefn56O4uBh2dnZ1vi5JEnx8fJCQkKDTe44ePRpubm7YvXs3AMDd3R3p6el45ZVXoFQqMWbM\nGEydOhVRUVFYtmwZ1q1bh8rKSnz66acICQnBqFGjMGvWLBQXFwMABg0ahLfeegv+/v7Yv38/MjMz\nMXXqVIwZMwZBQUH48ssvAVT/UhkyZAjmz5+PsWPHYtiwYdi/fz+A6l8QCxYswPDhwxEQEIDZs2ej\nsrISALBq1SqEhIQgODgYb7zxBjIzM+ts13//+18EBAQgKCgI06ZNQ1ZWFhITE+Hj46PZlkqlgp+f\nH+Lj41FUVIRZs2Zp6vz000+hVCoBAF5eXpg+fTqGDx+OqKioBvfp7Z/RnfaBUqnEnDlzEBQUhDFj\nxmDatGkoKSlBSkoKevXqBQAoLi7WvHdoaKjW5zpo0CCtempOnz17Fs8//zxCQkLw9NNP4/DhwwCA\nrKwsvPzyywgJCUFISAiWLl3aYHtI/9j90ox4eHjgypUrmuk///wTwcHBqKioQH5+Pjp16oSXX34Z\n/v7+da5fUFCAffv2YdCgQTq/p7u7u9Z7mpmZ4auvvkJQUBB27doFAAgNDcULL7yAESNGYMWKFTAy\nMsL27dshSRKWLFmCxYsXY+7cuQCArl27asJi4sSJeOmllzBo0CBUVFRg8uTJcHFxQY8ePZCcnAxf\nX198+OGH+PXXX/Hpp59i6NCh+OGHHxATE4Ndu3bB1NQUM2bMwN69ewEAV65cwZYtW2BsbIzNmzfj\ngw8+wJo1a7Tas23bNvzxxx/YunUrLC0t8cUXX2DWrFn4+uuv0bVrVxw6dAgjRozAsWPH4OzsDDc3\nN7z33nvw9PTEwoULoVKpMGvWLKxduxaTJ09GVVUVBg4ciGXLltW5/xr6jP71r3/VuQ8cHBxw5swZ\n7N27F5Ik4bPPPkNcXBwcHBw0216+fDnMzc3xyy+/IC8vDyEhIfD29q738ywoKMB7772Hr7/+Gs7O\nzsjMzMQzzzwDd3d37NixA87Ozvjmm29QWlqK2bNno6ioCDY2Njr/vVDjMdSbEUmSYG5urpnu3bs3\nVq9eDbVajZUrV2LPnj0YMWKE1jqLFi3CqlWrIP4eTWLgwIGYOHHiPb9nQ37//XcUFRXhxIkTAICq\nqiq0atVKq2YAKC0tRUREBAoKCjSBWFpaitjYWPTo0QMmJib4xz/+AQB45JFHkJ+fDwA4ceIEgoOD\nNTXd/IKYPn06oqKi8PTTTwOoPg9RVlZWq76jR49izJgxsLS0BFD9xfLll1+isrISY8eOxY4dOzBi\nxAhs374dzzzzjKZNUVFR2Lp1KwCgvLxca5s321SX+j6j+vaBr68vjIyMMG7cOPj6+mL48OHo0aMH\nUlJSNNs+efIk3n//fUiSBHt7ewwdOrS+jwYAcP78eWRlZWHq1KmaeZIkIS4uDn5+fpgyZQrS09PR\nv39/vP322wx0A2CoNyNRUVGak6c1KRQKvPnmmzh37hxmzZql+QkPADNnzqwV9Hf7njeDUhdqtRrv\nv/++JpBLSkpQUVGhef1mmKrVagghsGnTJs1J29zcXJiZmSEvLw8mJiZQKKp7FyVJ0qxvbKz9J5+d\nnQ21Wg21Wo1XXnkFzz//PACgsrISBQUFddZXc3tqtVrTleLv74+FCxciPj4eERERWLhwoWaZZcuW\nwc3NDQBQWFiotY2bbapPXZ9RffvAysoKu3btwtmzZ3Hq1Cm89dZbmDRpkma/3iRqDP1kZGR0x9dq\ndiu5ublhy5YtmtcyMzNhb28PExMTHDx4ECdPnsSpU6cwbtw4rFmzBl5eXg22j/SHferNRGJiIlau\nXImXX375jsvMmTMHx48fx4EDB/Tynlu2bEFKSsodu3Pq4uvriw0bNqCyshJqtRoffvghlixZUms5\na2tr9OzZE2vXrgVQHZTjx4/HwYMH691+v379sGfPHs32586di59//hm+vr7YunWrpv9+2bJlmDlz\nZq31/fz8sG3bNpSWlgIA1q9fjz59+sDU1BRmZmYYOXIkZs2ahWHDhmmC1tfXF+vWrYMQApWVlXj9\n9dfx/fff67xPaqr5GdW3Dw4fPoyXXnoJvXr1QlhYGEaPHo3o6Ohabdm6dSvUajUKCgq09p29vb1m\n+dOnTyMrKwsA0LNnT1y/fh0REREAgMuXL2P48OHIzMzE4sWLsXLlSgwZMgSzZ89Gly5d8Ndff91T\nO+ne8UhdpsrLyxEcHAyg+ijPzMwMM2bMwFNPPXXHdVxcXDB58mQsWLAAfn5+d/2ee/fuRWRkJCRJ\nglqthqurK7777ru7uvTujTfewL///W+EhIRApVKhe/fumDVrVp3LLl68GPPnz0dQUBAqKysRGBiI\nUaNGaXUx3O65555DamoqxowZAyEEnnjiCYSGhkKhUGj6hyVJgpOTk+ZIu6axY8ciPT0d48aNg1qt\nRseOHbF48WLN6+PGjcP333+vOQcAALNnz0Z4eDiCgoJQVVWF/v3745VXXtF5n9R0+2d0p32gUqlw\n9OhRBAYGwtLSEnZ2dpg/f77WtsLCwjBnzhz4+/vD3t5e61fcO++8g7lz52Lz5s3w9PSEp6cngOqw\nX758ORYtWoSKigoIIbBo0SI4OzvjxRdfxKxZsxAYGAhTU1O4u7tj5MiR99ROuneSEBx6l4hILtj9\nQkQkIwx1IiIZYagTEckIQ52ISEYY6kREMsJQJyKSEYY6EZGMMNSJiGSEoU5EJCMMdSIiGWGoExHJ\nCEOdiEhGGOpERDLCUCcikhGGOhGRjDDUiYhkhKFORCQjDHUiIhlhqBMRycgDFeoxMTGGLqHJsc3N\nA9ssfw9Kex+oUC8vLzd0CU2ObW4e2Gb5e1Da+0CFOhERNQ5DnYhIRhjqREQywlAnIpIRhjoRkYww\n1ImIZIShTkQkIwx1IiIZYagTEckIQ52ISEYY6kREMsJQJyKSEYY6EZGMMNSJiGSEoU5EJCMMdSIi\nGWGoExHJCEOdiEhGGOpERDJibOgCqHn6448/kJ2dbegymkxRURGSkpIMXYbOSktLAQCWlpb3tH7r\n1q3veV1qHIY6GUR2djays7PRyt7e0KU0CWsrKwi12tBl6Ky0pAQAYGFuftfr5uTmAgBcXFz0WhPp\nhqFOBtPK3h6BAf6GLoPqsGfvPgC4p8/n5rpkGOxTJyKSEYY6EZGMMNSJiGSEoU5EJCMMdSIiGWGo\nExHJCEOdiEhGGOpERDLCUCcikhGGOhGRjDDUiYhkhKFORCQjDHUiIhlhqBMRyQhDnYhIRhjqREQy\nwlAnIpIRhjoRkYww1ImIZIShTkQkIwx1IiIZYagTEckIQ52ISEYY6kREMsJQJyKSEYY6EZGMMNSJ\niGSEoU5EJCMMdSIiGWGoExHJCEOdiEhGGOpERDLCUCcikhGGOhGRjDDUiYhkhKFORCQjDHUiIhlh\nqBMRyQhDnYhIRhjqREQy8tCHemxsLGJjYw1dBhFRg5oir4zv69abwOXLlwEAHh4eBq6EiKh+TZFX\nD/2ROhER3cJQJyKSEYY6EZGMMNSJiGSEoU5EelVcAVxIBk7HFaOkrMrQ5TQ7D/3VL0T04EjPKcWu\n80aoVEnA9XxcTDqK5W8/BVMTI0OX1mzwSJ2I9OZgZGp1oP8tNasYEZcyDVhR88NQJyK9UUhS7XmK\n2vPo/mGoE5HeDO3THubGQjPduZ0dendva8CKmh/2qROR3rRpYYHBHirEZyvQxcUeL4Y8CRNjHjs2\nJe5tItKbHw/F4+doY8RmKHDwQiHSs0sMXVKzw1AnIr0oKKnE1t8TNdNlFWps/C3OgBU1Twx1ItKL\nkrIqqNRCa15+UYWBqmm+GOpEpBftWlvBvYOd1rzBfVwMVE3zxVAnIr15f2Iv9GivRqdWAmOftMfw\nvh0NXVKzw6tfiEhvIi5nIS5TQqUSsEouQ0WVCma8m7RJ8UidiPQit7Acq3ZeQoVSgoCEmKQy7Dxy\n1dBlNTsMdSLSi+sZxbVOlManFBiomuaLoU5EetG1gx3MTbW7Wnp0aW2gapovhjoR6YW1hQnefeEx\ntLISsDQV8H3EBgH9XQ1dVrPDUCcivahSqhEVnwuVABxsAe8uVhzMywAY6kSkF5sPxWP70WvIL5Vw\nLVvCxiPZhi6pWWKoE5FeRMZph/iNAiUycjj2S1NjqBORXrRvY6U1bW4qoaWtuYGqab4Y6kSkF6HD\nusKlrTUAwNRYIOiJlrzxyAB4RykRNdofF9Kx849rUEjAE51U8GgnobOLpaHLapYY6kTUKPGphVi2\nNRri7/uOrkGBtnai/pXovmH3CxE1yoWrOZpAryYhNc9Q1RBDnYgapaOjda159lZ1LEhNgqFORI3y\neLfWGNnPBcZGEoyNJHR3VKODvaGrar7Yp05EGociU7H96DUUFxvBs50agTqsE5WQixv5ZXi8aysE\n9HPB9bgzkCQeLxoKQ52IAABXUwvw3x2X/p6ScCrRCDGJufB0vfNh97X0Inzy7TnN6Izn43MR/Chg\nwwtfDIZfp0QEAIhOqH12M6qOeTWdjMnUGm63skqNpDyO92JIDHUiAgC4tbOtY55Nveu0aWFRa561\nmd5KonvAUCe9ySkoQ3FZlaHLoHvUyckGwb4dYW5qBCOFgFc7Nfp0d6h3nQGPOaKH263umX6eDnBu\nyWvUDalZ96mrVGpsO3wVf17OhIujDV4Y7sGxKu5BWYUSC7+NwNm4GzAxVuDZId3w7FD3OpctKK7A\n4YsFyMiW0PkBfH6CWi0gABg9QEPGpmSV4PK1PHRxtoOrU/1HzvdCrRb46qfLOHg2DQAwqJcTnIyT\nYKzDHf4KhYQBjzmhTQtz9HCzh28PJ+zZm6Z5XaVSo6JKBUtzE73XTXVr1qG+8bc4bD5wBQBw+Vou\nrqUVYvH0AQauqvFu5JXij3OpsLQwwT96tb/v/1B7jiXgbNwNANVjan//Syz6PeoEF0ftn/MqlRrv\nrjiG1KxiABKuZBjBJz4Xj7o9GNe/7Tp2DVt/T4RSpYa/TweEDu8KSTJsuB89n44vtkXjZrf1/wW4\nI7C/i17fIyI2C/v/TNVMH4hMw2APCS72DR9xL90SjZPRmQCAQ2fTYKS49eM/+noplv70G/KLK/C4\nuwP+NcEb1pameq2damvW3S8notK0puOS8pCdX2agavQjNasYYYsPY93Pl7By6wW8u+IYqpTq+/qe\nyZlFteal3CiuNS8mMefvQK8mIOFgZGqt5QzhSnIBvvvlL5SWK1FZpcauY9dxKuaGocvC5sPxqPnY\nzy2H46FW67d7I7mOzyq/tOH1cgrLNYEOAEIAP59MAgBUKIFdp/KQX1wBADgbdwOb9l/RT8FUr2Yd\n6o6ttG97szI3ho3Vw30k8cvJaygtV2qmr6UX4lzc/Q2nJzwdtaYtzIzwaB3PprSq4xeDlcWD8bP8\nah0PSL6aWmiASrRVVKq0piur1BBCv6H+eLfWqPmDRCEB7XXoFzdSSLi9l8rYqDpSCsqAKpX2NhJS\n+RDqpqBTqCuVSqxatQr+/v4ICAjA8OHD8eWXX2r9cR0/fhwvvvjifSv0fnhp5CNo07L67L2ZqRFe\nHdPjoR8qVFFHd8H97kHwfaw9Xn+6B7o426FXtzaYN6U/bOr4me3m3AIDerbXTFuYCAQ9qd+uhHv1\nSKeWtfbTI51aGKaYGvx9tPfP0D7OMDLS77FY53a2ePvZHujqbIeuznaY8VwPtLrDbf4qtcCG/Vcx\ndckxLP0xGk/UOJFqbCShZ9dWiLyuQGEZYGmmXWfPbm30WjfVTac+9Y8//hjZ2dnYvHkzbG1tUVxc\njKlTp8LGxgbjx4/HunXrsHr1anTr1u1+16vlzKUM7DwnoVIJKG2vYNxg3d5fpRZYv/cSjpxLhb2t\nGcYPdUf/Hu0MctSYVVCFD788gaTMQnh7tMXk0Y/CwuzuTnXEJORg3Z4YZOeXwbt7W1hbmGiuQunc\n3g693G/94x2/kIZN++NQUanCSF9XBA9w09pW1NVsrPs5BrmFFRjo7YwJI7o3+JzJrLwynI29gZyC\ncrRrbQ2n1rUTYf/p69j++1UAgO8jNjBVF6CdnRqO9rXvUjkUmYqdx65DCIHRfp0w2Lt9rWV0ceDP\nFOw6dh0KSULIgE54qle7Wsv8cjoZe04kwchIwhDv9rgYn4sqlRoj+7nA291wIVSzrsD+LlALgS7t\n7eDXw7HhlW9TVqHENz/H4dxf2XBpa4NJI921HmhxI68Mh8+lISu/DF6d7eHl2hJHrld3p/x4KB6H\nzqbB1soULwztgsT0Imw/kggAyMgtg62lCd6b0BM38stQXFqF9b/+hZvHip4dzAATS2TmlsL3sXbw\n79cJyzefQ2RsJlwcbfFqyKNwdtA+8bvv5DXsOnIVCoUC4wZ3xUDvDve8D8srlPjfT9GIuJSBDm1t\nMHn0o+joWPuyzaZQXFqJ1TuicDpKQitroN+Aklo9BfrSYHpkZGTgp59+wtGjR2FrW71DrK2t8dFH\nH+Hq1auIj49HfHw85s+fj/Xr19+XIuuSnV+GBevOQKmqDpzv9l6Go70V/Ho1HAC7/4jHtsNXNdtJ\nzYqBb897C47GEEJg09Ec5BRVd5fsP5MEUxMjvDamh87bKKtQYv43p1Hyd4j/euo6xg/rBisLE2Tk\nlMLS3ARXk/Ph0ckeaVnFWPT9n5o+2f/tika71lbo80h1UJSUVWH+N6dQVlH9k3/Lwb/QytYcI307\n11vD5z9EIiYhBwBw9HwqVGqBWS/20bweez0Xy388r5lOuQEEPgbU9aPoSnLNuxqBlTsuoYODFbp1\nuLuj5svX8rBq52XN9IrtMejgYA239rf+qaPic7Fmd6xmOjWrBJ9P7YuOjvq/wuRu1FXX4ql90eke\n6/p23xUc+vvKlryiHHy28QKWTuuveX3pj1GIS67uGjl2MQNCCHSzAf66IeF4fAIAICu/HAs3nEeX\n9tqhWFhaBVsrE/T2aIOwpce1XotNLcPmTwdrfv2u3HoB+89U97nnFmbh03VnsHLm4BrtzsbKrRc0\n0//ZeBYdHW3Rub3dPbX7u32X8eup63+/XwXC157B6lmDDXLye/WOKPx+NgWAhJQ84LPv/8Tn0/9x\nX96rwVC/ePEi3NzcYGenvWPd3Nzg5lZ9lBceHo7Tp0/fcRuFhYUoLKzdP2lra6v5orhb0Qk5UN7W\nZ7dzfwSyr51pcN3fYiQAtz7YkrIqrP1hN5ya+Nd2cTmQU6T9E/XY2QS0leJ13kZ6AVBSpr2NPyLi\nYGsOXEqvbuOPB67Ar6saKgGo1drLbtt3Cilx1fsxNQ8oq9B+fe+Ri6jMuoA7UamBmATtdSJiUrFj\nR4pm+nwycHtP39XUUlh1qN1vGxWfW2vexau5dx3qUQna2xGiel7NUL8Yn1PHMnkGD/W66opOyL3n\nUL942z5NvlGCvKIKtLQxQ0WlShPoNZfv1hNIK9AOv8oqNSzNtSPD1FgBp1bVv7ZqXwYqsPunn3Cz\nt+j4Oe3/u+TMYmzYtAOWf9+sFHld+3UhgA07DsHrHo+3/jirvb307BKs37QTNvfpquWioiIkJSXV\n+dqpKO1ariTlo7S86r5cmabT7/ya32y//PILVq1aBbVaDVNTU2zbtq3B9b/99lusWLGi1vw333wT\nYWFhd1HuLZ3ruPutlbVuJ5DsrYCUGnc/KySBFgYYKtTSFDA3VqNceSvw7nbI0haWgJEkoBK3PqMW\nFsCldO3lolIl+HWtvX/srW7Na2kJSJKAqLGthvapkQJoYSmQX1pzHe1l6vqVaW6swpUbZjgVk4k+\nHm00/cSd67iD0bWBuxrrUtf13LfPc63jb+h+XAd+t+r6225MXZ3b2SAz79ZVXfa2ZrD9+4IAM1Mj\ntGttibTsW5e7VL9XGVpZCSTWeJa0QgKeG+SGiko1ohJyYW1hjH6ebTHnm0goJAlervZIySrRjK3u\n3qYCRopb51bsrYHC8lvbszQVMK9x6qWVlUDN4Kued8/NRisr7at4zE0EDHVFZStraI0x79Ta6q67\nWXXV4Fa9vLwQHx+P4uJiWFtbY8SIERgxYgRSUlIwceJEnd7kxRdfREhISK3593qUDgAujraYHOyF\ndXuioFQBA3u7IOyZnpqz7/UZXl6FJT+cxZlLGbC1MsUro7zwVCP67hpDbX4c+86VICuvDN072ePd\nib3Ryq72rdf1aeeWjDW7olFUWoknHnHEG2N7YNInB6BW3bqUsYWdHaZMHIiWB69gy8ErqFKqMeSJ\njnhtTA+tI6y2nZPwze4YlJRVwsfLCTPGPw7zBv74vHrn4fMNkUjNKkHndnaYObE32rfRTnbbPTHY\n/Uf1T3kPZzNEJwko1RLObbwIb/fWeD+0FwCgV7fWGO3XCXv/vjTOv2+He+rb7u3RBqOe7Ih9p5Mh\nSUBgPxc81qWV1jL9PB0Q0LcDfotIgZFCQrBvJ3i6trzr99K3vp4OCOjXAb+dqa5rlG+negfVashL\n/u7ILaxAXHIBHFqYY+oYT63PfNpYLyzbEo30nFK4OtlgcpAHzp65ge5OAma2bXEyOhOW5sZ4YWhX\nuLazxdyXvVFYUonkGyX46Os/Ndu5llGEt8Z54fiZC7C3ktD3sS7w9vbWvO6XX4ZF6//E5Wu5cGhp\ngenP9UKPLrc+WyEErH+Kxr4T1yBJEkL+4YYJ/t3vud3/KCzHovV/IiYhB61bWGDaMz21zi/pW2Rk\npFZ7a+rrV4JF6yNwNaUAtuYCM55//L51A0lCh+ujPvroI+Tk5GDBggWwtbWFUqnEpk2bsG7dOhw4\ncAAAcPr0aaxYsaJR/er17ZQ72bptB4QAxo2t/aXRkLIKJUyNFXq/muBuREZGolevx1FeqWzUTzGV\nSo1KpVrz7f/N7hjs+PvEpCQBM8Y/rvniqlKqoVKrYW5ad1jfvi1dCCFQWq6s92RzZVV1X/3cVQcR\ndV37foAlb2r3Zd9c1rSRVyPpsp2KKhUUkgQT4wfrCl9911VaroS5qVGdJ76FENh2JBF/XMyAjYUJ\nOlrnwMlOIDDAH+WVqr/HSteu48dD8dh8KEFr3jMDO+PE+atIyZPQytYE/wrtC8/O2l+mpeVVMDc1\nvuMJ+Op2Aya63NKqg4beT190ya/NW3bAxAgYM+bu80pXOv3Xzp07F2vXrsXEiROhUqlQUlICHx8f\nrFmz5r4VpqvG5PH9+vlztxQKqdF9a0ZGCljU2BkvB3miZ9c2SEgrQM9ubdDF+VaftImxAib1XM16\n+7Z0IUlSg1cP3QzWuu6dUd02s7FhfjfbeVAvY9V3Xbf3h9d07GIGNh64dS7nL4UC47yrvxDNTeuu\no65zD1eSC5CSV/23k1OoxL+/i8A3Hw7T+kJo6G9d/+1+MO6FAIA7HEfplU5voVAoMGnSJEyaNOmO\ny/j4+MDHx0dvhVHjPe7hgMc97t/PzXv1RDdrXE4uhfrvvntP15Z19iNT0zn3l/bJWaVaQkZh/Ue2\nT3RvgxE+HXDgz+qT4v59O+DCVe2TsnlFFcjKK6vzMle6Px6MQ1VqVjo6mCG4p0Bithr9vB/FgMfu\n/tpr0i+XtrWfM9rSsv6eWUmSMDnIA50crXH2SjbMTIzQxdkWSZm3hh1o3cICDnXci0D3D0OdDKKl\nFdDCQmBI76a/P4Bq8+/bAVeSC3Dm8g2YmRjB1b4SKXkSYq/nw6PjnS8n3XsyCV//HAcAOHM5Cx4d\nW6Bz6+oHZTjYmeKdiX0eqBEvmwOGOhHBzMQIM59/DEWllTgelam5+enMmghMGNYFIQNc61zv93Pa\ng+LFXs/HOG81nvJQwMWlLbp2MPzVRM3Ng3W6n4gMysbSFLv+uKY1b/vRa3ccGdLutsccmRorcIfz\nqtREGOpEpEV5W4CrVGrcqXf9ucFusLao/sEvScCzg92a5AoPujOGOhFpCbrtIRwB/Vzu2C/u1t4W\nK2f44umnXBHs2xG9PTgSo6HxO5WItIzy7YTUa7G4niMh4KkeGPCYU73L/29PLI5eyAAA7DmRhMHu\nEto/GA+zapZ4pE5EWhLSCnEiQYG/shRYvesyTsZk3nHZrPwyTaADgFIlEJ3Gq10MiaFORFq+3XcF\npZXVwVxeqcKan2KhUtX9SMS6BhnR84OZ6C4x1IlIS0au9rg8BSWVKL/tsXo3ObS0QD/PW3ctKxQS\nPNsx1Q2JfepEpKWvpwP2nLg1Lvijne3rHdfnn888Cp/oTFzLKEJFlRqpyddhe3cDjZIeMdSJSMuE\nYV2RknQNaQUSenXvgOcGu9W7vJGRAr092mDD/qvIyi8HoMCldAFHp6qmKZi0sPuFiLSYGCvQy0WN\nkY+qMGVUd80DNeoTGZf9d5C5XtQAABGJSURBVKBXU6olnE8ouZ9l0h0w1Imo0cxMakfJgzY+fXPB\nvU5EjfZ4t9Zw73DrOcZWpgLeXTjcriGwT52IGs3ISIF5r/TGmcs38PPvF9DKCjA15vXqhsBQJyK9\nUKkEthxORFJmdQfApcxMeD1aqVOfPOkPu1+ISC9OxmRqPSAjr1iFw5HJBqyoeWKoE5FeVCpr33V6\n88Hf1HQY6kSkF/0826Klza2uFnNTCU893sGAFTVPDHUi0gsbSxN89kZf9GivQndHgVdHtEWblry1\ntKnxRCkR6c3hs2mITlNALSQUnciFT59K2FjyRGlT4pE6EelFZm4ZfjhwFWpRfSljSnYlth++auCq\nmh+GOhHpRUZuaa1hd1OziutemO4bhjoR6YWHSwvYWmqP5tjXy9FA1TRfDHUi0gszUyPMedkbLvZq\ntLERCOjdAoN6uzS8IukVQ52I9KaFdfVJ0cIy4EpqOW7klhq4ouaHoU5EevPlzstIylWgQinhano5\nPv8h0tAlNTsMdSLSm+jEXK3pS4m5d3y+Kd0fDHUi0hu3drZa067tbGFkxJhpStzbRKQ3rwZ3R2ur\n6usaHeyM8c/xjxu4ouaHd5QSkd60a22FoMdUUAkFOrs6wrWdXcMrkV7xSJ2I9M7YyNAVNF8MdSIi\nGWGoExHJCEOdiEhGGOpERDLCUCcikpGH/pLG7t27G7oEIiKdNEVePfSh7uHhYegSiIh00hR5xe4X\nIiIZYagTEckIQ52ISEYY6kREMsJQJyKSEYY6EZGMMNSJiGSEoU5EJCMMdSIiGWGoExHJCEOdiEhG\nGOpERDLCUCcikhGGOhGRjDDUiYhkhKFORCQjDHUiIhlhqBMRyQhDnYhIRhjqREQywlAnIpIRhjoR\nkYww1ImIZIShTkQkIwx1IiIZYagTEckIQ52ISEYY6kREMsJQJyKSEYY6EZGMMNSJiGSEoU5EJCMM\ndSIiGWGoExHJCEOdiEhGGOpERDLCUCcikhGGOhGRjDDUiYhkhKFORCQjDHUiIhkxNnQB1Hzl5OZi\nz959hi6D6pCTkwMA9/T55OTmonXr1vouiXTEUCeDaG7/9EVFRbCxsTF0GTqztLICAEiKu/8x37p1\n62b3+T5IGOpkEH5+foYuoUlFRkbC29vb0GU0qcjISEOX0CyxT52ISEYY6kREMsJQJyKSEYY6EZGM\nMNSJiGSEoU5EJCMMdSIiGWGoExHJCEOdiEhGGOpERDLCUCcikhGGOhGRjDDUiYhkhKFORCQjDHUi\nIhlhqBMRyQhDnYhIRhjqREQy8sA8zk6pVCIrKwspKSmGLqVJsc3NA9ssf1lZWVAqlTA2NmysSkII\nYdAK/hYbG4vg4GBDl0FEdM927doFDw8Pg9bwwBypW1hYAAA2bNgAR0dHA1fTNDIyMvDCCy+wzTLH\nNsu/zTfbezPHDOmBCXUjIyMAgKOjI5ydnQ1cTdNim5sHtln+buaYIfFEKRGRjDDUiYhkhKFORCQj\nRnPnzp1r6CJuMjMzg4+PD8zMzAxdSpNhm5sHtln+HpT2PjCXNBIRUeOx+4WISEYY6kREMmLQUE9L\nS8MLL7yAESNG4PXXX0dJSUmtZSorK/H2228jKCgIwcHBOHHihAEq1R9d2/zJJ59g9OjRGDlyJI4d\nO2aASvVDl/beVFxcjCFDhuD06dNNWKH+6dLmGzduYNKkSQgODkZISAhOnjxpgEobb/fu3QgICMCw\nYcOwYcOGWq9fvnwZY8aMwfDhwzF79mwolUoDVKlfDbX5wIEDCA4OxqhRo/DGG2+goKCgaQsUBjRl\nyhSxZ88eIYQQK1asEIsWLaq1zI8//ijeeustIYQQsbGxws/Pr0lr1Ddd2vzf//5XzJgxQ6jVanHl\nyhXh6+sr1Gp1U5eqF7q096aZM2eKPn36iFOnTjVVefeFLm1+++23xffffy+EECI+Pl70799fKJXK\nJq2zsTIyMsTAgQNFXl6eKCkpEUFBQeKvv/7SWmbkyJHi3LlzQggh3nvvPbFhwwZDlKo3DbW5qKhI\nPPnkkyIjI0MIIcTSpUvF/Pnzm7RGgx2pV1VVISIiAsOHDwcAjBkzBr/88kut5dRqNcrKyqBSqVBW\nVgZzc/OmLlVvdG3zvn37MHnyZEiShK5du2Lt2rUQD+H5bF3bCwB79+6FlZUV3N3dm7JEvdO1zUOH\nDkVgYCAAoGPHjqioqEBpaWmT1tpYJ06cQN++fdGiRQtYWlpi+PDhWm1NTU1FeXk5evbsCaD+z/9h\n0VCbq6qqMGfOHLRt2xYA4O7ujvT09Cat0WChnpeXB2tra82IZm3atEFmZmat5UJCQpCfnw8/Pz9M\nmDAB77zzTlOXqje6tvn69euIiIjA888/j2effRbZ2dlQKB6+0x+6tjctLQ3ffvstZs6c2dQl6p2u\nbR4+fDjs7OwAAF9//TW6d+8OGxubJq21sW7cuIE2bdpoph0cHLTaevvrd9oXD5OG2tyyZUsMHToU\nAFBeXo6vvvoKQ4YMadIam2Tsl3379mHBggVa8zp27AhJkrTm3T4NACtWrEDPnj2xceNGXLt2DS+9\n9BI8PT3Rvn37+1pzYzWmzSqVChkZGdiwYQPi4uLwyiuvYN++fQ/0P/29tletVmP27Nn48MMPH7pf\nYY35jG9at24dNm/ejO+///6+1Hg/qdVqrbYJIbSmG3r9YaRrm4qKijB16lR4eHggJCSkKUtsmlD3\n9/eHv7+/1ryqqir4+PhApVLByMgIWVlZcHBwqLXuwYMH8Z///AeSJMHV1RWPPfYYLl68+MCHemPa\n3Lp1a4wcORKSJMHDwwOOjo5ITExEjx49mqr8u3av7U1ISEBCQgJmz54NAEhKSsIHH3yA+fPno2/f\nvk1W/71ozGcMAIsWLcKRI0ce2pEMHR0d8eeff2qmb2+ro6MjsrKyNNPZ2dl33BcPi4baDNw6Cd63\nb1+8//77TV2i4bpfTExM0Lt3b+zduxcAsHPnTgwYMKDWch4eHjhw4AAAIDc3F9HR0ejevXuT1qov\nurZ54MCBmmWSk5ORnp4OV1fXJq1VH3Rpb5cuXXDkyBHs2rULu3btgpeXFz755JMHPtDvRNfPeN26\ndTh9+jQ2btz4UAY6APTv3x8nT55Ebm4uysrK8Ntvv2m1tX379jAzM0NkZCSA6rHG69oXD5OG2qxS\nqfDaa6/B398fs2fPNsgvE4PeUZqamopZs2YhJycHTk5OWLJkCezs7LBx40bcuHED06dPR3Z2Nj78\n8EMkJSVBoVDg1Vdf1Zxgehjp0ubi4mLMmzcPMTExAIB33nkHAwcONHDl90aX9tYUGhqKN998Ez4+\nPgaquPEaavO0adPwxBNPwNraGra2tpr1vvrqK80JtofF7t27sXr1alRVVWHs2LGYPHkyJk+ejGnT\npuHRRx9FbGwsPvjgAxQXF8PT0xMLFiyAqampoctulPranJGRgbCwMK0T/l5eXggPD2+y+jhMABGR\njDx8l1QQEdEdMdSJiGSEoU5EJCMMdSIiGWGoExHJSJPcfER0u3nz5qFly5YICwvD5MmT8e6776JL\nly6N3u4XX3yBDRs2aC4NVCqVcHd3x7Rp09CpUycA1ZdNpqamau7QVavVqKysxOuvv47Ro0cjJSUF\nQ4cORbdu3TTbFUJg4sSJGDt2bL3vn5ubi379+iEuLg4HDx7EyZMn8cEHH+Dy5csICwuDra0tli5d\nigULFiAhIQGhoaGYMGFCo9tNdBNDnQxuzZo1et1eQEAAPvroI830zp078eKLL+Lnn3+GtbU1AGDm\nzJkYMWKEZpmoqCiMHz9eM06Hubk5du3apXk9MzMTgYGB8PLygoeHh051DB48GIMHDwZQfWe0j48P\nwsPDkZaWhmPHjuH8+fMwMjJqdHuJamKoU4NOnz6NJUuWwMnJCYmJibCwsMCUKVOwfv16JCYmYtiw\nYZrboQ8dOoRVq1ahqqoK5ubmePfdd9GrVy8UFxdj9uzZiI2NhYODA4yMjODt7Q0AGDRoEJYtWwZP\nT098+umnuHDhAkpKSiCEwCeffAJvb2/MmjUL1tbWiIuLQ0ZGBtzd3fHvf/8bVlZWDdY/evRo/PTT\nT9i9ezfGjx9f5zLJycmwtLS8440xbdu2RceOHXHt2rVaof7bb7/hP//5DywsLODl5aWZv337dvz6\n668YOXIkNm7cCJVKhcLCQsTHx0OpVGLMmDH44osvUFVVhfDwcOTn50OlUiE0NBRjx47F6dOnER4e\nDktLS5SUlGDbtm04duxYnfv3iy++QGpqKrKyspCamoq2bdvis88+g4ODAxITE/HRRx8hNzcXCoUC\nr7/+OgICApCZmYl58+YhPT0dVVVVGDlyJF577TWd/iboAdakA/3SQ+nUqVOie/fuIiYmRgghxKRJ\nk8Szzz4rKioqRE5OjvD09BQZGRkiMTFRBAYGitzcXCGEEFeuXBFPPvmkKCkpEeHh4WLmzJlCrVaL\nnJwcMWDAALF8+XIhhBADBw4UFy9eFGfPnhVhYWFCpVIJIYRYvXq1ePXVV4UQQrz77rua96ysrBSj\nR48WW7durVXr8uXLxccff1xr/sKFC8XcuXOFEEJMmDBBDBw4UIwaNUo89dRTol+/fuKf//ynpn3J\nycmiZ8+eWuufPXtW9OnTR6SlpWnNz8rKEt7e3poxtb/88kvRrVs3IYQQ27ZtE1OmTKlVV83tV1VV\niYCAABEdHS2EEKKwsFD4+/uLc+fOiVOnTgkPDw+RkpIihBD17t/ly5eLwYMHi6KiIiGEEK+++qpY\ntmyZEEKI0aNHa8ZuT0tL0ywXGhoqDh48KIQQory8XISGhoqff/65zr8BenjwSJ104uzsjEceeQQA\n4OLiAhsbG5iamsLe3h5WVlYoKChAREQEbty4gZdeekmzniRJSEpKwsmTJ/H+++9DkiTY29trhiet\nqVevXrCzs8OmTZuQnJyM06dPax2J+/n5aY6ku3XrdldPlJEkSWsUyJvdL7m5uZg8eTLatm2raR9Q\nPWxqcHAwgOrxPFq2bInPPvsMTk5OWtuNjIxEt27dNOcDnn32WSxZskTnuq5du4akpCStgZ/Ky8tx\n6dIluLm5wcnJSTN43fHjx++4fwFohh4AgEceeQQFBQXIz89HbGwsxo0bBwBwcnLCgQMHUFpaioiI\nCBQUFGDZsmUAgNLSUsTGxiIgIEDn+unBw1AnndzeLXFzvPCa1Go1+vXrh6VLl2rmpaena0axEzVG\npKirL/n3339HeHg4/u///g+DBw9G586d8dNPP2lerxnKkiTd1YNDoqKi8PTTT9eab29vj6VLlyIw\nMBC9evXCsGHDNO9Vs0+9PjXrqGu/1EelUsHGxkbrvbKzs2FjY4Pz58/D0tJSM7++/bt///4698/N\nemoOLJWQkIA2bdpACIFNmzbBwsICQPVJXjMzs7uqnx48vKSR9KZfv344fvw44uPjAQBHjhzBqFGj\nUF5eDj8/P2zduhVqtRoFBQU4ePBgrfWPHz+OgQMH4vnnn4eXlxcOHDgAlUrV6Lq2bNmClJSUWsPk\n3tShQwe89tprCA8Pv+unD/Xp0wdXr15FbGwsgOp+9Lvh6uqq9QWSnp6OwMBAREdH11q2vv17J9bW\n1vD09MTOnTs12x8/frzmiURr164FABQWFmL8+PF1fi70cOGROulNly5dMG/ePMyYMUNzlLhq1SpY\nWVkhLCwMc+bMgb+/P+zt7bUuF7zpueee0zxkXKlU4sknn8Rvv/0GtVp9V3Xs3bsXkZGRkCQJarUa\nrq6u+O677+o9Cp00aRJ27tyJVatW4dlnn9X5vezt7bF48WK88847MDExQZ8+fe6qVlNTU6xcuRLh\n4eH43//+B6VSienTp8Pb27vWA7jr27/1+fzzz/Hxxx9j/fr1kCQJ4eHhaNOmDRYvXoz58+cjKCgI\nlZWVCAwMxKhRo+6qfnrwcJRGIiIZYfcLEZGMMNSJiGSEoU5EJCMMdSIiGWGoExHJCEOdiEhGGOpE\nRDLCUCcikpH/BxhQhvVoT3iOAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Identification of all sequences that satisfy the cutoff and contain the specified motif\n",
    "sequences = []\n",
    "for i in range(seq_cutoff):\n",
    "    if degron_seq_scores[i][0].find(motif) == index:\n",
    "        sequences.append(degron_seq_scores[i][0])\n",
    "        \n",
    "drp_medians = []\n",
    "for s in sequences:\n",
    "    mutated_drp = mutate(sequence = s, index = index)\n",
    "    for i in range(kmer_length):\n",
    "        drp_medians.append([motif[i]+str(i+1), np.median(mutated_drp[i])])\n",
    "drp_medians = pd.DataFrame(drp_medians, columns = ['residue', 'median DRP difference'])\n",
    "\n",
    "sns.set(style=\"ticks\")\n",
    "\n",
    "# Initialize the figure\n",
    "f, ax = plt.subplots(figsize=(6, 4))\n",
    "\n",
    "# Create boxplots\n",
    "sns.boxplot(x=\"median DRP difference\", y=\"residue\", data=drp_medians,\n",
    "            palette=\"vlag\", fliersize=0)\n",
    "\n",
    "# Add in points to show each observation\n",
    "sns.swarmplot(x=\"median DRP difference\", y=\"residue\", data=drp_medians)\n",
    "\n",
    "# Tweak the visual presentation\n",
    "ax.xaxis.grid(True)\n",
    "ax.set(ylabel=\"\")\n",
    "sns.despine(trim=True, left=True)\n",
    "plt.title('DRP Difference over Residues')\n",
    "plt.savefig(wd+'figures/mutations/boxplot_'+motif,dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mutating two positions at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "string index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-39-e4e023cb23b7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m plot_weights_double(weights = np.transpose(np.asarray(mutated_drp)),\n\u001b[1;32m      7\u001b[0m     \u001b[0mplot_title\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'Motif = '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mmotif\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m', Mutations at Positions '\u001b[0m\u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmut_index1\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' and '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmut_index2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m                     sequence = sequence, index1 = mut_index1, index2 = mut_index2)\n\u001b[0m",
      "\u001b[0;32m<ipython-input-34-c98933675f16>\u001b[0m in \u001b[0;36mplot_weights_double\u001b[0;34m(weights, plot_title, sequence, index1, index2)\u001b[0m\n\u001b[1;32m    132\u001b[0m     \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_xticklabels\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maminoacids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    133\u001b[0m     \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_yticklabels\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maminoacids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 134\u001b[0;31m     \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_xlabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Alternate AA ('\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' Position '\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex1\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m')'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    135\u001b[0m     \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_ylabel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Alternate AA ('\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex2\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m' Position '\u001b[0m \u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex2\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m')'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    136\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: string index out of range"
     ]
    }
   ],
   "source": [
    "mut_index1 = index + 1\n",
    "mut_index2 = index + 2\n",
    "mutated_drp = mutate_double(sequence = sequence, index1 = mut_index1, index2 = mut_index2)\n",
    "\n",
    "# Plot DRP differences using a heatmap and save the figure\n",
    "plot_weights_double(weights = np.transpose(np.asarray(mutated_drp)),\n",
    "    plot_title = 'Motif = ' + motif + ', Mutations at Positions '+ str(mut_index1) + ' and ' + str(mut_index2), \n",
    "                    sequence = sequence, index1 = mut_index1, index2 = mut_index2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mutating three positions at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mut_index1 = index\n",
    "mut_index2 = index + 1\n",
    "mut_index3 = index + 2\n",
    "mutated_drp_3d = mutate_triple(sequence = sequence, index1 = mut_index1, index2 = mut_index2, index3 = mut_index3)\n",
    "\n",
    "# # Plot DRP differences using a heatmap and save the figure\n",
    "# plot_weights_double(weights = np.transpose(np.asarray(mutated_drp)),\n",
    "#     plot_title = 'Motif = ' + motif + ', Mutations at Positions '+ str(mut_index1) + ' and ' + str(mut_index2), \n",
    "#                     sequence = sequence, index1 = mut_index1, index2 = mut_index2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "flattened_drp = mutated_drp_3d.flatten()\n",
    "percentage = 0.01\n",
    "\n",
    "high_indices = []\n",
    "for i in range(int(len(flattened_drp) * percentage)):\n",
    "    max_index = np.argmax(flattened_drp)\n",
    "    # Calculate 3 dimensions based on flat index\n",
    "    dim1 = int(max_index/400)\n",
    "    dim2 = int((max_index - dim1*400)/20)\n",
    "    dim3 = int(max_index-dim1*400-dim2*20)\n",
    "    high_indices.append([dim1, dim2, dim3, flattened_drp[max_index]]) \n",
    "    # create list of info with 3 dimensions and the DRP change\n",
    "    flattened_drp[max_index] = -100  \n",
    "    \n",
    "flattened_drp = mutated_drp_3d.flatten()\n",
    "low_indices = []\n",
    "for i in range(int(len(flattened_drp) * percentage)):\n",
    "    min_index = np.argmin(flattened_drp)\n",
    "    dim1 = int(min_index/400)\n",
    "    dim2 = int((min_index - dim1*400)/20)\n",
    "    dim3 = int(min_index-dim1*400-dim2*20)\n",
    "    low_indices.append([dim1, dim2, dim3, flattened_drp[min_index]]) \n",
    "    flattened_drp[min_index] = 100  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from matplotlib.pyplot import figure\n",
    "\n",
    "\n",
    "# Make 3D scatterplots for mutations with highest degron gain and loss\n",
    "\n",
    "# Degron Gain\n",
    " \n",
    "df_gain = pd.DataFrame(high_indices, columns = ('X', 'Y', 'Z', 'DRP change'))\n",
    "fig = plt.figure(figsize=(12,8))\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "scatter = ax.scatter(df_gain['X'], df_gain['Y'], df_gain['Z'], c=list(df_gain['DRP change']), s=60, cmap = 'viridis')\n",
    "ax.view_init(30, 200)\n",
    "ax.set_xlabel('Alternate AA (' + sequence[mut_index1] + ' Position ' +str(mut_index1) + ')', labelpad = 5)\n",
    "ax.set_ylabel('Alternate AA (' + sequence[mut_index2] + ' Position ' +str(mut_index2) + ')', labelpad = 5)\n",
    "ax.set_zlabel('Alternate AA (' + sequence[mut_index3] + ' Position ' +str(mut_index3) + ')', labelpad = 5)\n",
    "#ax.set_xticklabels(aminoacids)\n",
    "#plt.xticks(np.arange(20), aminoacids)\n",
    "# plt.yticks(np.arange(20), aminoacids)\n",
    "#plt.zticks(np.arange(20), aminoacids)\n",
    "# ax.set_yticklabels(aminoacids)\n",
    "#ax.set_zticklabels(np.arange(20), aminoacids)\n",
    "plt.title('Degron Gain Mutations for ' + motif + ' Motif', fontsize = 14, pad = 18)\n",
    "fig.colorbar(scatter, shrink=0.8)\n",
    "\n",
    "# ax.annotate(\"HII\", xyz=(5, 5, 5), \n",
    "#                   xyztext=(3, 10, 5), #va=\"bottom\", ha=\"center\",\n",
    "#                   bbox=dict(boxstyle=\"round\", fc=\"w\"),\n",
    "#                   arrowprops=dict(arrowstyle=\"->\"))\n",
    "\n",
    "plt.savefig(wd+'figures/mutations/3d_gain_'+motif,dpi=300)\n",
    "plt.close()\n",
    "\n",
    "\n",
    "# Degron Loss \n",
    "df_loss = pd.DataFrame(low_indices, columns = ('X', 'Y', 'Z', 'DRP change'))\n",
    "fig = plt.figure(figsize=(12,8))\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "scatter = ax.scatter(df_loss['X'], df_loss['Y'], df_loss['Z'], c=list(df_loss['DRP change']), s=60, cmap = 'viridis')\n",
    "ax.view_init(30, 200)\n",
    "ax.set_xlabel('Alternate AA (' + sequence[mut_index1] + ' Position ' +str(mut_index1) + ')', labelpad = 5)\n",
    "ax.set_ylabel('Alternate AA (' + sequence[mut_index2] + ' Position ' +str(mut_index2) + ')', labelpad = 5)\n",
    "ax.set_zlabel('Alternate AA (' + sequence[mut_index3] + ' Position ' +str(mut_index3) + ')', labelpad = 5)\n",
    "#ax.set_xticklabels(aminoacids)\n",
    "#plt.xticks(np.arange(20), aminoacids)\n",
    "# plt.yticks(np.arange(20), aminoacids)\n",
    "#plt.zticks(np.arange(20), aminoacids)\n",
    "# ax.set_yticklabels(aminoacids)\n",
    "#ax.set_zticklabels(np.arange(20), aminoacids)\n",
    "plt.title('Degron Loss Mutations for ' + motif + ' Motif', fontsize = 14, pad = 18)\n",
    "fig.colorbar(scatter, shrink=0.8)\n",
    "\n",
    "# ax.annotate(\"HII\", xyz=(5, 5, 5), \n",
    "#                   xyztext=(3, 10, 5), #va=\"bottom\", ha=\"center\",\n",
    "#                   bbox=dict(boxstyle=\"round\", fc=\"w\"),\n",
    "#                   arrowprops=dict(arrowstyle=\"->\"))\n",
    "\n",
    "plt.savefig(wd+'figures/mutations/3d_loss_'+motif,dpi=300)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "top_gain_mutations = []\n",
    "for i in range(5): \n",
    "    mutation = aminoacids[high_indices[i][0]] + aminoacids[high_indices[i][1]] + aminoacids[high_indices[i][2]]\n",
    "    top_gain_mutations.append([mutation, df_gain[0:5]['DRP change'][i]])\n",
    "top_gain_mutations = pd.DataFrame(top_gain_mutations, columns = ['Mutation', 'DRP Gain'])\n",
    "\n",
    "top_loss_mutations = []\n",
    "for i in range(5):\n",
    "    mutation = aminoacids[low_indices[i][0]] + aminoacids[low_indices[i][1]] + aminoacids[low_indices[i][2]]\n",
    "    top_loss_mutations.append([mutation, df_loss[0:5]['DRP change'][i]])\n",
    "top_loss_mutations = pd.DataFrame(top_loss_mutations, columns = ['Mutation', 'DRP Loss'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "top_gain_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "top_loss_mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Weighting algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A(-1)': [852,\n",
       "  0.05241141732283464,\n",
       "  217,\n",
       "  2.4217539120732743e-27,\n",
       "  9.6870156482931e-26],\n",
       " 'A(-2)': [958,\n",
       "  0.05893208661417323,\n",
       "  221,\n",
       "  1.6784254229557e-22,\n",
       "  4.0282210150936795e-21],\n",
       " 'A(-3)': [1018,\n",
       "  0.06262303149606299,\n",
       "  155,\n",
       "  0.00024249659308052778,\n",
       "  0.00181872444810396],\n",
       " 'A(-4)': [938,\n",
       "  0.05770177165354331,\n",
       "  117,\n",
       "  0.18009943808056406,\n",
       "  0.568735067622834],\n",
       " 'A(-5)': [1045,\n",
       "  0.06428395669291338,\n",
       "  142,\n",
       "  0.021081871607808923,\n",
       "  0.0872353307909335],\n",
       " 'A(-6)': [992, 0.0610236220472441, 101, 0.8977378919581022, 1.0],\n",
       " 'C(-1)': [516,\n",
       "  0.03174212598425197,\n",
       "  101,\n",
       "  2.654060869454089e-07,\n",
       "  2.65406086945409e-06],\n",
       " 'C(-2)': [454,\n",
       "  0.02792814960629921,\n",
       "  104,\n",
       "  6.452104504342176e-11,\n",
       "  8.602806005789571e-10],\n",
       " 'C(-3)': [385, 0.023683562992125984, 40, 0.7522137207262336, 1.0],\n",
       " 'C(-4)': [487,\n",
       "  0.02995816929133858,\n",
       "  61,\n",
       "  0.25391888466977214,\n",
       "  0.7254825276279201],\n",
       " 'C(-5)': [384, 0.023622047244094488, 44, 0.5169403753314259, 1.0],\n",
       " 'C(-6)': [338, 0.02079232283464567, 31, 0.9117336553004571, 1.0],\n",
       " 'D(-1)': [805, 0.04952017716535433, 46, 0.9999999790268913, 1.0],\n",
       " 'D(-2)': [813, 0.0500123031496063, 28, 0.9999999999999998, 1.0],\n",
       " 'D(-3)': [745, 0.04582923228346457, 47, 0.9999984774364252, 1.0],\n",
       " 'D(-4)': [746, 0.045890748031496065, 56, 0.9997741351967382, 1.0],\n",
       " 'D(-5)': [694, 0.04269192913385827, 47, 0.9999756085862236, 1.0],\n",
       " 'D(-6)': [725, 0.04459891732283464, 44, 0.9999993530603987, 1.0],\n",
       " 'E(-1)': [805,\n",
       "  0.04952017716535433,\n",
       "  96,\n",
       "  0.35328826889984283,\n",
       "  0.942102050399581],\n",
       " 'E(-2)': [1151,\n",
       "  0.07080462598425197,\n",
       "  165,\n",
       "  0.0020224172927866134,\n",
       "  0.0127731618491786],\n",
       " 'E(-3)': [1071, 0.06588336614173228, 93, 0.9982119026904701, 1.0],\n",
       " 'E(-4)': [1258,\n",
       "  0.07738681102362205,\n",
       "  166,\n",
       "  0.03269571683673998,\n",
       "  0.126564065174477],\n",
       " 'E(-5)': [1148, 0.07062007874015748, 126, 0.6986055158316333, 1.0],\n",
       " 'E(-6)': [1178, 0.07246555118110236, 129, 0.7100127020079711, 1.0],\n",
       " 'F(-1)': [738, 0.0453986220472441, 71, 0.9428516192361994, 1.0],\n",
       " 'F(-2)': [579, 0.03561761811023622, 24, 0.9999999995469131, 1.0],\n",
       " 'F(-3)': [611, 0.0375861220472441, 47, 0.9987089793856048, 1.0],\n",
       " 'F(-4)': [568, 0.034940944881889764, 35, 0.9999864848883999, 1.0],\n",
       " 'F(-5)': [640, 0.03937007874015748, 38, 0.9999984724907958, 1.0],\n",
       " 'F(-6)': [595, 0.036601870078740155, 52, 0.9828519167193236, 1.0],\n",
       " 'G(-1)': [559,\n",
       "  0.0343873031496063,\n",
       "  177,\n",
       "  8.813554129213008e-33,\n",
       "  1.0576264955055598e-30],\n",
       " 'G(-2)': [953,\n",
       "  0.05862450787401575,\n",
       "  135,\n",
       "  0.007281109044201414,\n",
       "  0.0379883950132248],\n",
       " 'G(-3)': [1006,\n",
       "  0.06188484251968504,\n",
       "  127,\n",
       "  0.1369590263558053,\n",
       "  0.48338479890284203],\n",
       " 'G(-4)': [1036, 0.06373031496062992, 73, 0.9999985280406029, 1.0],\n",
       " 'G(-5)': [1006, 0.06188484251968504, 94, 0.9834444625554547, 1.0],\n",
       " 'G(-6)': [960, 0.05905511811023622, 113, 0.39172412253123207, 1.0],\n",
       " 'H(-1)': [544, 0.03346456692913386, 52, 0.9201363574764779, 1.0],\n",
       " 'H(-2)': [459, 0.02823572834645669, 23, 0.999998739443693, 1.0],\n",
       " 'H(-3)': [446, 0.027436023622047244, 27, 0.9999288908791856, 1.0],\n",
       " 'H(-4)': [436, 0.026820866141732284, 36, 0.9842704665478564, 1.0],\n",
       " 'H(-5)': [408, 0.025098425196850394, 48, 0.44266902527903534, 1.0],\n",
       " 'H(-6)': [514,\n",
       "  0.031619094488188976,\n",
       "  63,\n",
       "  0.30697465108853256,\n",
       "  0.837203593877816],\n",
       " 'I(-1)': [743, 0.04570620078740158, 39, 0.9999999954370173, 1.0],\n",
       " 'I(-2)': [642, 0.03949311023622047, 67, 0.7946570017886632, 1.0],\n",
       " 'I(-3)': [592,\n",
       "  0.03641732283464567,\n",
       "  79,\n",
       "  0.09356750640357465,\n",
       "  0.35087814901340497],\n",
       " 'I(-4)': [636,\n",
       "  0.039124015748031496,\n",
       "  95,\n",
       "  0.00611488014286757,\n",
       "  0.0333538916883686],\n",
       " 'I(-5)': [617,\n",
       "  0.03795521653543307,\n",
       "  121,\n",
       "  1.5539267711051717e-08,\n",
       "  1.6951928412056402e-07],\n",
       " 'I(-6)': [664, 0.04084645669291338, 77, 0.4683422918631144, 1.0],\n",
       " 'K(-1)': [1218, 0.07492618110236221, 59, 0.9999999999999992, 1.0],\n",
       " 'K(-2)': [1263, 0.07769438976377953, 93, 0.9999991552995645, 1.0],\n",
       " 'K(-3)': [1238, 0.07615649606299213, 66, 0.9999999999999173, 1.0],\n",
       " 'K(-4)': [1238, 0.07615649606299213, 94, 0.9999957879143677, 1.0],\n",
       " 'K(-5)': [1271, 0.0781865157480315, 78, 0.9999999999148066, 1.0],\n",
       " 'K(-6)': [1206, 0.07418799212598425, 126, 0.8661044514426495, 1.0],\n",
       " 'L(-1)': [1822, 0.11208169291338582, 65, 0.9999999999999999, 1.0],\n",
       " 'L(-2)': [1411, 0.08679872047244094, 111, 0.9999949792285746, 1.0],\n",
       " 'L(-3)': [1455,\n",
       "  0.08950541338582677,\n",
       "  179,\n",
       "  0.16428883799785374,\n",
       "  0.563276015992641],\n",
       " 'L(-4)': [1428,\n",
       "  0.08784448818897637,\n",
       "  200,\n",
       "  0.001997860981311826,\n",
       "  0.0127731618491786],\n",
       " 'L(-5)': [1424,\n",
       "  0.08759842519685039,\n",
       "  262,\n",
       "  3.1330584370726993e-14,\n",
       "  6.2661168741454e-13],\n",
       " 'L(-6)': [1469,\n",
       "  0.09036663385826772,\n",
       "  177,\n",
       "  0.24612745283872806,\n",
       "  0.7203730326987159],\n",
       " 'M(-1)': [381, 0.0234375, 35, 0.9222395889416108, 1.0],\n",
       " 'M(-2)': [284, 0.017470472440944882, 26, 0.895494729698088, 1.0],\n",
       " 'M(-3)': [336, 0.02066929133858268, 31, 0.9058380670412808, 1.0],\n",
       " 'M(-4)': [301, 0.018516240157480313, 23, 0.9846705977975116, 1.0],\n",
       " 'M(-5)': [342,\n",
       "  0.021038385826771654,\n",
       "  74,\n",
       "  3.169674260692912e-07,\n",
       "  2.92585316371653e-06],\n",
       " 'M(-6)': [333,\n",
       "  0.020484744094488187,\n",
       "  53,\n",
       "  0.01205503206047668,\n",
       "  0.0535779202687852],\n",
       " 'N(-1)': [662, 0.04072342519685039, 64, 0.9275592816354379, 1.0],\n",
       " 'N(-2)': [607, 0.03734005905511811, 38, 0.9999893580136247, 1.0],\n",
       " 'N(-3)': [549, 0.03377214566929134, 45, 0.9929810990335765, 1.0],\n",
       " 'N(-4)': [530, 0.03260334645669291, 41, 0.9972216139411494, 1.0],\n",
       " 'N(-5)': [574, 0.03531003937007874, 50, 0.9822180823714476, 1.0],\n",
       " 'N(-6)': [589, 0.03623277559055118, 54, 0.9613390576602124, 1.0],\n",
       " 'P(-1)': [886, 0.05450295275590551, 77, 0.9957879748070755, 1.0],\n",
       " 'P(-2)': [947, 0.05825541338582677, 65, 0.9999984318255205, 1.0],\n",
       " 'P(-3)': [1019, 0.06268454724409449, 117, 0.4979896067839546, 1.0],\n",
       " 'P(-4)': [1021,\n",
       "  0.06280757874015748,\n",
       "  127,\n",
       "  0.17692842510425122,\n",
       "  0.568735067622834],\n",
       " 'P(-5)': [1045,\n",
       "  0.06428395669291338,\n",
       "  128,\n",
       "  0.22484055007083303,\n",
       "  0.691817077141025],\n",
       " 'P(-6)': [1064,\n",
       "  0.0654527559055118,\n",
       "  147,\n",
       "  0.011703533733494264,\n",
       "  0.0535779202687852],\n",
       " 'Q(-1)': [793, 0.048781988188976375, 42, 0.9999999980596759, 1.0],\n",
       " 'Q(-2)': [791, 0.04865895669291338, 85, 0.7381886054249573, 1.0],\n",
       " 'Q(-3)': [731, 0.044968011811023625, 72, 0.9153243458574293, 1.0],\n",
       " 'Q(-4)': [802,\n",
       "  0.049335629921259845,\n",
       "  115,\n",
       "  0.00906063884145374,\n",
       "  0.0434910664389779],\n",
       " 'Q(-5)': [802, 0.049335629921259845, 83, 0.8394011085044412, 1.0],\n",
       " 'Q(-6)': [824,\n",
       "  0.050688976377952756,\n",
       "  118,\n",
       "  0.008590189878601865,\n",
       "  0.0429509493930093],\n",
       " 'R(-1)': [828,\n",
       "  0.05093503937007874,\n",
       "  125,\n",
       "  0.0012819405748261852,\n",
       "  0.00904899229289072],\n",
       " 'R(-2)': [1041,\n",
       "  0.0640378937007874,\n",
       "  123,\n",
       "  0.3694453509506486,\n",
       "  0.9637704807408219],\n",
       " 'R(-3)': [848,\n",
       "  0.05216535433070866,\n",
       "  212,\n",
       "  1.0215755324468177e-25,\n",
       "  3.06472659734045e-24],\n",
       " 'R(-4)': [977,\n",
       "  0.06010088582677165,\n",
       "  118,\n",
       "  0.28531627762087863,\n",
       "  0.796231472430359],\n",
       " 'R(-5)': [1076, 0.06619094488188976, 71, 0.9999999406371788, 1.0],\n",
       " 'R(-6)': [1022,\n",
       "  0.06286909448818898,\n",
       "  161,\n",
       "  3.666017574446054e-05,\n",
       "  0.000293281405955684],\n",
       " 'S(-1)': [1624,\n",
       "  0.09990157480314961,\n",
       "  222,\n",
       "  0.003503865367662205,\n",
       "  0.0210231922059732],\n",
       " 'S(-2)': [1543, 0.09491879921259842, 160, 0.9127827728819671, 1.0],\n",
       " 'S(-3)': [1640, 0.10088582677165354, 181, 0.7064695512103816, 1.0],\n",
       " 'S(-4)': [1432, 0.08809055118110236, 155, 0.7764172409591863, 1.0],\n",
       " 'S(-5)': [1440, 0.08858267716535433, 157, 0.7480099819583532, 1.0],\n",
       " 'S(-6)': [1424, 0.08759842519685039, 156, 0.7265858901627846, 1.0],\n",
       " 'T(-1)': [734,\n",
       "  0.04515255905511811,\n",
       "  103,\n",
       "  0.021813933290628303,\n",
       "  0.0872557331625132],\n",
       " 'T(-2)': [861,\n",
       "  0.05296505905511811,\n",
       "  125,\n",
       "  0.0046041347364411825,\n",
       "  0.0263093413510925],\n",
       " 'T(-3)': [1096,\n",
       "  0.06742125984251969,\n",
       "  139,\n",
       "  0.1140269182245891,\n",
       "  0.4146433389985061],\n",
       " 'T(-4)': [933,\n",
       "  0.05739419291338583,\n",
       "  130,\n",
       "  0.013474357860979416,\n",
       "  0.0577472479756261],\n",
       " 'T(-5)': [862,\n",
       "  0.05302657480314961,\n",
       "  108,\n",
       "  0.17864713266963503,\n",
       "  0.568735067622834],\n",
       " 'T(-6)': [853,\n",
       "  0.05247293307086614,\n",
       "  105,\n",
       "  0.23418853410274373,\n",
       "  0.702565602308231],\n",
       " 'V(-1)': [1075,\n",
       "  0.06612942913385826,\n",
       "  200,\n",
       "  2.0684754719025175e-11,\n",
       "  3.10271320785378e-10],\n",
       " 'V(-2)': [782,\n",
       "  0.04810531496062992,\n",
       "  210,\n",
       "  2.4819899849446165e-29,\n",
       "  1.4891939909667698e-27],\n",
       " 'V(-3)': [812,\n",
       "  0.049950787401574805,\n",
       "  151,\n",
       "  7.452891468267266e-09,\n",
       "  8.94346976192072e-08],\n",
       " 'V(-4)': [822,\n",
       "  0.050565944881889764,\n",
       "  168,\n",
       "  8.776223382124262e-13,\n",
       "  1.50449543693559e-11],\n",
       " 'V(-5)': [855,\n",
       "  0.05259596456692913,\n",
       "  143,\n",
       "  6.13541998415351e-06,\n",
       "  5.25893141498872e-05],\n",
       " 'V(-6)': [876, 0.05388779527559055, 89, 0.8873154443944916, 1.0],\n",
       " 'W(-1)': [186, 0.011441929133858268, 17, 0.852682886198907, 1.0],\n",
       " 'W(-2)': [255, 0.015686515748031496, 23, 0.8972797879788037, 1.0],\n",
       " 'W(-3)': [224, 0.013779527559055118, 9, 0.9999554437499297, 1.0],\n",
       " 'W(-4)': [238, 0.014640748031496063, 16, 0.9924308365504141, 1.0],\n",
       " 'W(-5)': [209, 0.012856791338582677, 11, 0.9989175599201066, 1.0],\n",
       " 'W(-6)': [201, 0.01236466535433071, 17, 0.9191764651199612, 1.0],\n",
       " 'Y(-1)': [485, 0.02983513779527559, 52, 0.7016247277102685, 1.0],\n",
       " 'Y(-2)': [462, 0.028420275590551183, 34, 0.9979425536501291, 1.0],\n",
       " 'Y(-3)': [434, 0.026697834645669292, 43, 0.8489335772466167, 1.0],\n",
       " 'Y(-4)': [429, 0.026390255905511813, 34, 0.9909872240833232, 1.0],\n",
       " 'Y(-5)': [414, 0.02546751968503937, 35, 0.9752524651349675, 1.0],\n",
       " 'Y(-6)': [429, 0.026390255905511813, 47, 0.6384095500954354, 1.0]}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kmer_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PPM1D gene analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# PPM1D gene analysis\n",
    "kmer_length = t\n",
    "motif = 'KTVCVC'\n",
    "index = 17\n",
    "sequence = 'RLRGQKKIGNPLLHQHRKTVCVC'\n",
    "\n",
    "mutated_drp = mutate(sequence = sequence, index = index)\n",
    "\n",
    "weights = np.transpose(np.asarray(mutated_drp))\n",
    "plot_title = 'PPM1D gene'\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "im = ax.imshow(weights, cmap='viridis')\n",
    "k = len(motif)\n",
    "\n",
    "# Set tick labels\n",
    "residues = list(motif)\n",
    "ax.set_xticks(np.arange(k))\n",
    "ax.set_yticks(np.arange(len(aminoacids)))\n",
    "ax.set_xticklabels(residues)\n",
    "ax.set_yticklabels(aminoacids);\n",
    "\n",
    "# Rotate the tick labels and set their alignment.\n",
    "# plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\", rotation_mode=\"anchor\")\n",
    "cbar = plt.colorbar(im)\n",
    "im.set_clim(-0.06, 0) # revise colorbar range\n",
    "\n",
    "# Turn spines off and create white grid\n",
    "for edge, spine in ax.spines.items():\n",
    "    spine.set_visible(False)\n",
    "ax.set_xticks(np.arange(weights.shape[1]+1)-.5, minor=True)\n",
    "ax.set_yticks(np.arange(weights.shape[0]+1)-.5, minor=True)\n",
    "ax.grid(which=\"minor\", color=\"w\", linestyle='-', linewidth=3)\n",
    "ax.tick_params(which=\"minor\", bottom=False, left=False)\n",
    "\n",
    "# set size of tick labels\n",
    "plt.tick_params(axis='both', which='major', labelsize=18)\n",
    "cbar.ax.tick_params(labelsize=18) \n",
    "\n",
    "# set title\n",
    "plt.title(plot_title, fontsize=20)\n",
    "\n",
    "from matplotlib import rcParams\n",
    "rcParams['axes.titlepad'] = 20 \n",
    "\n",
    "plt.savefig(wd+'figures/mutations/'+'PPM1D',dpi=300)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Position analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "positions = []\n",
    "for row in top_motifs:\n",
    "    kmer = row[0]\n",
    "    #positions = {}\n",
    "    positions_one_kmer = []\n",
    "    for seq in top_sequences:\n",
    "        location = seq[0].find(kmer)\n",
    "        if location > -1:  # if kmer exists in sequence\n",
    "            #positions.update({location : positions[location]+1})\n",
    "            positions_one_kmer.append(location)\n",
    "    positions.append(positions_one_kmer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plt.close()\n",
    "sns.distplot(positions[0])\n",
    "plt.title(\"Position Frequency, Motif = CGC\")\n",
    "plt.savefig(wd+'figures/motifs/CGC',dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "positions[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "top_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "kmer_info_sorted[0][1][3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Old code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Calculate baseline frequency by averaging frequencies\n",
    "# How likely a kmer is to appear at random \n",
    "\n",
    "# freq_sum = 0\n",
    "# for i in kmer_dict:\n",
    "#     freq_sum += kmer_dict[i]\n",
    "# baseline = freq_sum / (num_seq * len(kmer_dict))\n",
    "\n",
    "# GENERAL / AVERAGE\n",
    "# Determine motif cutoff using Binomial expression and baseline\n",
    "\n",
    "# i = 0\n",
    "# while stats.binom.sf(i, seq_cutoff, baseline) > 0.05:\n",
    "#     i = i+1\n",
    "# motif_cutoff = i\n",
    "# binom_value = stats.binom.sf(i, seq_cutoff, baseline)\n",
    "\n",
    "# def binomcdf_orig():\n",
    "#     #p = baseline\n",
    "#     p = baseline_prob\n",
    "#     n = seq_cutoff\n",
    "#     x = 100\n",
    "#     result = []\n",
    "#     labels = []\n",
    "#     for a in range(20):\n",
    "#         labels.append(x)\n",
    "#         result.append(stats.binom.sf(x, n, p))\n",
    "#         x += 5\n",
    "#     return result, labels\n",
    "\n",
    "# # plt.hist(binomcdf())\n",
    "# # plt.show()\n",
    "# b, l = binomcdf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# # Calculates motif-specific p-value\n",
    "# # Lower p-value = higher enrichment of kmer among top sequences\n",
    "# def p_value(kmer):\n",
    "#     baseline_prob = kmer_dict[kmer] / num_seq\n",
    "    \n",
    "#     # Frequency kmer among top sequences\n",
    "#     top_freq = kmer_dict_top[kmer] \n",
    "    \n",
    "#     return stats.binom.sf(top_freq, seq_cutoff, baseline_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#     lines = [[(middle - 100, p_value), (top_count, p_value)], \n",
    "#              [(top_count, -0.05), (top_count, p_value)]]\n",
    "#     c = np.array([(1, 0, 0, 1), (0, 1, 0, 1), (0, 0, 1, 1)])\n",
    "\n",
    "#     lc = mc.LineCollection(lines, colors=c, linewidths=2)\n",
    "#     fig, ax = pl.subplots()\n",
    "#     ax.add_collection(lc)\n",
    "#     ax.autoscale()\n",
    "#     ax.margins(0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Seaborn stuff\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sns.set(style=\"ticks\")\n",
    "\n",
    "rs = np.random.RandomState(11)\n",
    "x = rs.gamma(2, size=1000)\n",
    "y = -.5 * x + rs.normal(size=1000)\n",
    "\n",
    "sns.jointplot(x, y, kind=\"hex\", color=\"#4CB391\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# p_vals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Load the example flights dataset and conver to long-form\n",
    "flights_long = sns.load_dataset(\"flights\")\n",
    "flights = flights_long.pivot(\"month\", \"year\", \"passengers\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Draw a heatmap with the numeric values in each cell\n",
    "f, ax = plt.subplots(figsize=(9, 6))\n",
    "sns.heatmap(flights, annot=True, fmt=\"d\", linewidths=.5, ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Motif analysis with DRP score weighting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Function description: \n",
    "#\n",
    "# Cutoff -- number of sequences from the top + bottom to consider for motif analysis\n",
    "# Motif length -- length of kmer motifs to be considered\n",
    "\n",
    "def motif_with_weighting(cutoff = 50, motif_length = 2):\n",
    "    top = degron_seq_scores[0:cutoff]\n",
    "    bottom = degron_seq_scores[-cutoff:]\n",
    "    kmers = []\n",
    "    \n",
    "    for row in top:\n",
    "        sequence = row[0]\n",
    "        for j in range(seq_length - motif_length + 1):\n",
    "            kmers.append((sequence[j:(j+motif_length)], row[1])) # add k-mer and corresponding score\n",
    "    \n",
    "    # Normalize drp scores to emphasize differences\n",
    "    kmers = np.asarray(kmers) # Convert kmers to np array from a list\n",
    "    x = np.asarray(list(map(float, kmers[:, 1])))\n",
    "    normalized = (x-min(x))/(max(x)-min(x))\n",
    "    kmers[:,1] = normalized\n",
    "    \n",
    "    # Convert list of kmers into 1 string\n",
    "    kmer_str = \" \".join(kmers[:, 0])\n",
    "    kmer_list = list(kmers[:, 0])\n",
    "    \n",
    "    unique_kmers = [] \n",
    "  \n",
    "    # Determine all unique kmers\n",
    "    for i in kmer_list: \n",
    "        # checking for duplicacy \n",
    "        if i not in unique_kmers: \n",
    "            # insert value in str2 \n",
    "            unique_kmers.append(i)  \n",
    "    \n",
    "    for i in range(0, len(unique_kmers)): \n",
    "        # \n",
    "        np.asarrray\n",
    "        kmer_dict.update({unique_kmers[i] : kmer_list.count(unique_kmers[i])})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
